---
title: "Sequencing read QC"
author:
  # - Jane Doe^[Institution Two, jane@example.org]      # add report authors (uncomment if using)
  # - John Doe^[Institution One, john@example.org]      # add a second report author (uncomment if using)
date: "Date: `r base::format(Sys.time(), '%d/%m/%Y')`"
always_allow_html: true
output:
  bookdown::html_document2:
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: true
    number_sections: false
    self_contained: yes
    css: ../esr-styles.css
    highlight: pygments
    includes:
      after_body: ../esr_footer.html
  editor_options: 
    chunk_output_type: console
---

<img src="../images/ESR_logo.svg" style="position:absolute;top:19px;right:40px;width:23%;" />

```{r setup, include=FALSE}
# setup default chunk settings
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, error = FALSE, message = FALSE, fig.align = "center")
```

## Info

Explore the number of raw sequencing reads and the number of aligned/unaligned reads for each pipeline run

```{r, out.width = "100%"}
# load libraries
library(dplyr)
library(tidyr)
library(plotly)
library(gtools)
library(DT)

# for icons at bottom of doc
htmltools::tagList(rmarkdown::html_dependency_font_awesome())

# read in yaml config file
config <- yaml::yaml.load_file("../config/config.yaml")

# read in metadata
metadata <- utils::read.csv(file.path(config$metadata))

# read in mapping rates info and remove empty rows
mapping_rates_data <- utils::read.csv("../mapping_rates/mapping_rates.csv") %>%
  tidyr::drop_na()

# double check the samples are in a consistant order - naturally sorted by sample names
mapping_rates_data <- mapping_rates_data[gtools::mixedorder(base::as.character(mapping_rates_data$sample)),]

# mixedsort of samples was overwritten by plotly so I needed to specifically define the natural sorting of the sample names here
sample_list <- gtools::mixedsort(mapping_rates_data$sample)

xform <- base::list(categoryorder = "array",
                    categoryarray = sample_list)

# flag samples with low read counts
mapping_rates_data <- mapping_rates_data %>%
  dplyr::mutate(low_fastq_read_count = dplyr::case_when(raw_read_count_fastq_file < config$low_sequencing_read_count ~ TRUE,
                                                        raw_read_count_fastq_file > config$low_sequencing_read_count ~ FALSE))

# join metadata with mapping rates data
mapping_rates_data <- dplyr::left_join(mapping_rates_data, metadata, by = "sample")

# create a vector to decide if the r chunks should be evaluated based on whether the dats from a given pipeline was analysed
smrnaseq_chunk_eval <- config$mirna_smrnaseq
excerpt_chunk_eval <- base::any("TRUE" %in% c(config$mirna_excerpt,
                                              config$pirna_excerpt,
                                              config$trna_excerpt,
                                              config$circrna_excerpt,
                                              config$gencode_excerpt))
```

## Alignment rates {.tabset .tabset-fade}

```{r, eval = smrnaseq_chunk_eval, results = 'asis'}
base::cat(base::paste0("### smrnaseq"))
```

```{r, eval = smrnaseq_chunk_eval, out.width = "100%"}
# create a bar chart of the raw read counts
mapping_rates_data %>%
  plotly::plot_ly(x = ~sample,
                  y = ~raw_read_count_fastq_file,
                  type = "bar",
                  color = I("#0097db"),
                  textposition = "none",
                  name = "Raw sequencing \n read count",
                  hoverinfo = "text",
                  text = ~base::paste("</br> Raw sequencing read count:", base::format(raw_read_count_fastq_file, big.mark = ",", scientific = FALSE, digits = 2),
                                      "</br> Sample:", sample,
                                      "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~smrnaseq_reads_mapped_to_mature_bowtie_stats_file,
                    color = I("#85C659"),
                    name = "Reads aligned to mature RNA's",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Reads aligned to mature RNA's:", base::format(smrnaseq_reads_mapped_to_mature_bowtie_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file,
                    color = I("#ec1515"),
                    name = "Reads aligned to hairpin RNA's",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Reads aligned to hairpin RNA's:", base::format(smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~smrnaseq_reads_unmapped_to_mature_bowtie_stats_file,
                    color = I("#febf2a"),
                    name = "Reads unaligned to mature RNA's",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Reads unaligned to mature RNA's:", base::format(smrnaseq_reads_unmapped_to_mature_bowtie_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file,
                    color = I("#784f96"),
                    name = "Reads unaligned to hairpin RNA's",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Reads unaligned to hairpin RNA's:", base::format(smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  # add line that show you the average raw read count across all samples
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$raw_read_count_fastq_file),
                       yend = base::mean(mapping_rates_data$raw_read_count_fastq_file),
                       name = "Average raw \n sequencing read count",
                       color = I("#0097db")) %>%
  # add line that show you the average number of reads aligned to mature RNA's
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$smrnaseq_reads_mapped_to_mature_bowtie_stats_file),
                       yend = base::mean(mapping_rates_data$smrnaseq_reads_mapped_to_mature_bowtie_stats_file),
                       name = "Average reads aligned to \n mature RNA's",
                       color = I("#85C659")) %>%
  # add line that show you the average number of reads aligned to hairpin RNA's
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file),
                       yend = base::mean(mapping_rates_data$smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file),
                       name = "Average reads aligned to \n hairpin RNA's",
                       color = I("#ec1515")) %>%
  # add line that show you the average number of reads unmapped to mature RNA's
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$smrnaseq_reads_unmapped_to_mature_bowtie_stats_file),
                       yend = base::mean(mapping_rates_data$smrnaseq_reads_unmapped_to_mature_bowtie_stats_file),
                       name = "Average reads unaligned to \n mature RNA's",
                       color = I("#febf2a")) %>%
  # add line that show you the average number of reads unmapped to hairpin RNA's
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file),
                       yend = base::mean(mapping_rates_data$smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file),
                       name = "Average reads unaligned to \n hairpin RNA's",
                       color = I("#784f96")) %>%
  plotly::layout(xaxis = xform) %>%
  plotly::layout(xaxis = list(title = "Sample"),
                 yaxis = list(title = "Read count"))
```

```{r, eval = smrnaseq_chunk_eval, out.width = "100%"}
# create datatable to explore the RNA's that are found in the media
DT::datatable(mapping_rates_data %>%
                dplyr::mutate(across(c("sample",
                                       "treatment",
                                       "low_fastq_read_count"), as.factor)) %>%
                dplyr::mutate(across(c("raw_read_count_fastq_file",
                                       "smrnaseq_reads_mapped_to_mature_bowtie_stats_file",
                                       "smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file",
                                       "smrnaseq_reads_unmapped_to_mature_bowtie_stats_file",
                                       "smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file"), as.integer)) %>%
                dplyr::select(sample,
                              treatment,
                              raw_read_count_fastq_file,
                              smrnaseq_reads_mapped_to_mature_bowtie_stats_file,
                              smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file,
                              smrnaseq_reads_unmapped_to_mature_bowtie_stats_file,
                              smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file,
                              low_fastq_read_count),
              filter = "top",
              rownames = FALSE,
              colnames = c("Sample",
                           "Treatment",
                           "Raw sequencing read count",
                           "Reads aligned to mature RNA's",
                           "Reads aligned to hairpin RNA's",
                           "Reads unaligned to mature RNA's",
                           "Reads unaligned to hairpin RNA's",
                           "Low raw sequencing read count"),
              extensions = list("ColReorder" = NULL,
                                "Buttons" = NULL,
                                "FixedColumns" = list(leftColumns=1)),
              options = list(
                dom = "BRrltpi",
                autoWidth = TRUE,
                columnDefs = list(list(width = "200px", targets = 0)),
                lengthMenu = list(c(10, 50, -1), c("10", "50", "All")),
                ColReorder = TRUE,
                buttons =
                  list("copy",
                       list(extend = "collection",
                            buttons = c("csv", "excel", "pdf"),
                            text = "Download"),
                       I("colvis"))))
```

```{r, eval = excerpt_chunk_eval, results = 'asis'}
base::cat(base::paste0("### excerpt"))
```

```{r, eval = excerpt_chunk_eval, out.width = "100%"}
# create a bar chart of the raw read counts
mapping_rates_data %>%
  plotly::plot_ly(x = ~sample,
                  y = ~raw_read_count_fastq_file,
                  type = "bar",
                  color = I("#0097db"),
                  textposition = "none",
                  name = "Raw sequencing \n read count",
                  hoverinfo = "text",
                  text = ~base::paste("</br> Raw sequencing read count:", base::format(raw_read_count_fastq_file, big.mark = ",", scientific = FALSE, digits = 2),
                                      "</br> Sample:", sample,
                                      "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~excerpt_reads_used_for_alignment_stats_file,
                    color = I("#85C659"),
                    name = "Reads used for alignment",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Reads used for alignment:", base::format(excerpt_reads_used_for_alignment_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~excerpt_not_mapped_to_genome_or_libs_stats_file,
                    color = I("#ec1515"),
                    name = "Not mapped to \n genome or libs",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Not mapped to \n genome or libs:", base::format(excerpt_not_mapped_to_genome_or_libs_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  # add line that show you the average raw read count across all samples
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$raw_read_count_fastq_file),
                       yend = base::mean(mapping_rates_data$raw_read_count_fastq_file),
                       name = "Average raw sequencing \n read count",
                       color = I("#0097db")) %>%
  # add line that show you the average number of reads used for alignment across all samples
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$excerpt_reads_used_for_alignment_stats_file),
                       yend = base::mean(mapping_rates_data$excerpt_reads_used_for_alignment_stats_file),
                       name = "Average reads used \n for alignment",
                       color = I("#85C659")) %>%
  # add line that show you the average number of reads not mapped to genome or libs
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$excerpt_not_mapped_to_genome_or_libs_stats_file),
                       yend = base::mean(mapping_rates_data$excerpt_not_mapped_to_genome_or_libs_stats_file),
                       name = "Average reads not \n mapped to genome or libs",
                       color = I("#ec1515")) %>%
  plotly::layout(xaxis = xform) %>%
  plotly::layout(xaxis = list(title = "Sample"),
                 yaxis = list(title = "Read count"))
```

```{r, eval = excerpt_chunk_eval, out.width = "100%"}
# create datatable to explore the RNA's that are found in the media
DT::datatable(mapping_rates_data %>%
                dplyr::mutate(across(c("sample",
                                       "treatment",
                                       "low_fastq_read_count"), as.factor)) %>%
                dplyr::mutate(across(c("raw_read_count_fastq_file",
                                       "excerpt_reads_used_for_alignment_stats_file",
                                       "excerpt_not_mapped_to_genome_or_libs_stats_file"), as.integer)) %>%
                dplyr::select(sample,
                              treatment,
                              raw_read_count_fastq_file,
                              excerpt_reads_used_for_alignment_stats_file,
                              excerpt_not_mapped_to_genome_or_libs_stats_file,
                              low_fastq_read_count),
              filter = "top",
              rownames = FALSE,
              colnames = c("Sample",
                           "Treatment",
                           "Raw sequencing read count",
                           "Reads used for alignment",
                           "Not mapped to genome or libs",
                           "Low raw sequencing read count"),
              extensions = list("ColReorder" = NULL,
                                "Buttons" = NULL,
                                "FixedColumns" = list(leftColumns=1)),
              options = list(
                dom = "BRrltpi",
                autoWidth = TRUE,
                columnDefs = list(list(width = "200px", targets = 0)),
                lengthMenu = list(c(10, 50, -1), c("10", "50", "All")),
                ColReorder = TRUE,
                buttons =
                  list("copy",
                       list(extend = "collection",
                            buttons = c("csv", "excel", "pdf"),
                            text = "Download"),
                       I("colvis"))))
```

```{r, cleanup, results = "hide"}
# clean up
rm(list = ls())
```
