---
title: "Sequencing read QC"
author:
  # - Jane Doe^[Institution Two, jane@example.org]      # add report authors (uncomment if using)
  # - John Doe^[Institution One, john@example.org]      # add a second report author (uncomment if using)
date: "Date: `r base::format(Sys.time(), '%d/%m/%Y')`"
output:
  html_document:
    code_download: true
  editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
# setup default chunk settings
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, error = FALSE, message = FALSE, fig.align = "center")
```

# Context

# Sequencing read counts {.tabset .tabset-fade}

## smrnaseq

```{r, out.width = "100%"}
# load libraries
library(dplyr)
library(tidyr)
library(plotly)
library(gtools)
library(DT)

# read in yaml config file
config <- yaml::yaml.load_file("../config/config.yaml")

# read in metadata
metadata <- utils::read.csv(file.path(config$metadata))

# read in mapping rates info and remove empty rows
mapping_rates_data <- utils::read.csv("../mapping_rates/mapping_rates.csv") %>%
  tidyr::drop_na()

# double check the samples are in a consistant order - naturally sorted by sample names
mapping_rates_data <- mapping_rates_data[gtools::mixedorder(base::as.character(mapping_rates_data$sample)),]

# mixedsort of samples was overwritten by plotly so I needed to specifically define the natural sorting of the sample names here
sample_list <- gtools::mixedsort(mapping_rates_data$sample)

xform <- base::list(categoryorder = "array",
                    categoryarray = sample_list)

# flag samples with low read counts
mapping_rates_data <- mapping_rates_data %>%
  dplyr::mutate(low_fastq_read_count = dplyr::case_when(raw_read_count_fastq_file < config$low_sequencing_read_count ~ TRUE,
                                                        raw_read_count_fastq_file > config$low_sequencing_read_count ~ FALSE))

# join metadata with mapping rates data
mapping_rates_data <- dplyr::left_join(mapping_rates_data, metadata, by = "sample")

# create a bar chart of the raw read counts
mapping_rates_data %>%
  plotly::plot_ly(x = ~sample,
                  y = ~raw_read_count_fastq_file,
                  type = "bar",
                  textposition = "none",
                  name = "Raw sequencing \n read count",
                  hoverinfo = "text",
                  text = ~base::paste("</br> Raw sequencing read count:", base::format(raw_read_count_fastq_file, big.mark = ",", scientific = FALSE, digits = 2),
                                      "</br> Sample:", sample,
                                      "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~smrnaseq_reads_mapped_to_mature_bowtie_stats_file,
                    name = "Reads aligned to mature RNA's",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Reads aligned to mature RNA's:", base::format(smrnaseq_reads_mapped_to_mature_bowtie_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file,
                    name = "Reads aligned to hairpin RNA's",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Reads aligned to hairpin RNA's:", base::format(smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~smrnaseq_reads_unmapped_to_mature_bowtie_stats_file,
                    name = "Reads unaligned to mature RNA's",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Reads unaligned to mature RNA's:", base::format(smrnaseq_reads_unmapped_to_mature_bowtie_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file,
                    name = "Reads unaligned to hairpin RNA's",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Reads unaligned to hairpin RNA's:", base::format(smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  # add line that show you the average raw read count across all samples
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$raw_read_count_fastq_file),
                       yend = base::mean(mapping_rates_data$raw_read_count_fastq_file),
                       name = "Average raw \n sequencing read count",
                       color = I("#1f77b4")) %>%
  # add line that show you the average number of reads aligned to mature RNA's
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$smrnaseq_reads_mapped_to_mature_bowtie_stats_file),
                       yend = base::mean(mapping_rates_data$smrnaseq_reads_mapped_to_mature_bowtie_stats_file),
                       name = "Average reads aligned to \n mature RNA's",
                       color = I("#ff7f0e")) %>%
  # add line that show you the average number of reads aligned to hairpin RNA's
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file),
                       yend = base::mean(mapping_rates_data$smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file),
                       name = "Average reads aligned to \n hairpin RNA's",
                       color = I("#2ca02c")) %>%
  # add line that show you the average number of reads unmapped to mature RNA's
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$smrnaseq_reads_unmapped_to_mature_bowtie_stats_file),
                       yend = base::mean(mapping_rates_data$smrnaseq_reads_unmapped_to_mature_bowtie_stats_file),
                       name = "Average reads unaligned to \n mature RNA's",
                       color = I("#d62728")) %>%
  # add line that show you the average number of reads unmapped to hairpin RNA's
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file),
                       yend = base::mean(mapping_rates_data$smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file),
                       name = "Average reads unaligned to \n hairpin RNA's",
                       color = I("#9467bd")) %>%
  plotly::layout(xaxis = xform) %>%
  plotly::layout(xaxis = list(title = "Sample"),
                 yaxis = list(title = "Read count"))
```

```{r, out.width = "100%"}
# create datatable to explore the RNA's that are found in the media
DT::datatable(mapping_rates_data %>%
                dplyr::mutate_if(is.numeric, format, big.mark = ",", scientific = FALSE, digits = 0) %>%
                dplyr::mutate(across(c("sample", "treatment", "low_fastq_read_count"), as.factor)) %>%
                dplyr::select(sample,
                              treatment,
                              raw_read_count_fastq_file,
                              smrnaseq_reads_mapped_to_mature_bowtie_stats_file,
                              smrnaseq_reads_mapped_to_hairpin_bowtie_stats_file,
                              smrnaseq_reads_unmapped_to_mature_bowtie_stats_file,
                              smrnaseq_reads_unmapped_to_hairpin_bowtie_stats_file,
                              low_fastq_read_count),
              filter = "top",
              rownames = FALSE,
              colnames = c("Sample",
                           "Treatment",
                           "Raw sequencing read count",
                           "Reads aligned to mature RNA's",
                           "Reads aligned to hairpin RNA's",
                           "Reads unaligned to mature RNA's",
                           "Reads unaligned to hairpin RNA's",
                           "Low raw sequencing read count"),
              extensions = list("ColReorder" = NULL,
                                "Buttons" = NULL,
                                "FixedColumns" = list(leftColumns=1)),
              options = list(
                dom = "BRrltpi",
                autoWidth = TRUE,
                columnDefs = list(list(width = "200px", targets = 0)),
                lengthMenu = list(c(10, 50, -1), c("10", "50", "All")),
                ColReorder = TRUE,
                buttons =
                  list("copy",
                       list(extend = "collection",
                            buttons = c("csv", "excel", "pdf"),
                            text = "Download"),
                       I("colvis"))))
```

## excerpt

```{r, out.width = "100%"}
# create a bar chart of the raw read counts
mapping_rates_data %>%
  plotly::plot_ly(x = ~sample,
                  y = ~raw_read_count_fastq_file,
                  type = "bar",
                  textposition = "none",
                  name = "Raw sequencing \n read count",
                  hoverinfo = "text",
                  text = ~base::paste("</br> Raw sequencing read count:", base::format(raw_read_count_fastq_file, big.mark = ",", scientific = FALSE, digits = 2),
                                      "</br> Sample:", sample,
                                      "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~excerpt_reads_used_for_alignment_stats_file,
                    name = "Reads used for alignment",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Reads used for alignment:", base::format(excerpt_reads_used_for_alignment_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~excerpt_not_mapped_to_genome_or_libs_stats_file,
                    name = "Not mapped to \n genome or libs",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Not mapped to \n genome or libs:", base::format(excerpt_not_mapped_to_genome_or_libs_stats_file, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  # add line that show you the average raw read count across all samples
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$raw_read_count_fastq_file),
                       yend = base::mean(mapping_rates_data$raw_read_count_fastq_file),
                       name = "Average raw sequencing \n read count",
                       color = I("#1f77b4")) %>%
  # add line that show you the average number of reads used for alignment across all samples
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$excerpt_reads_used_for_alignment_stats_file),
                       yend = base::mean(mapping_rates_data$excerpt_reads_used_for_alignment_stats_file),
                       name = "Average reads used \n for alignment",
                       color = I("#ff7f0e")) %>%
  # add line that show you the average number of reads not mapped to genome or libs
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(mapping_rates_data$excerpt_not_mapped_to_genome_or_libs_stats_file),
                       yend = base::mean(mapping_rates_data$excerpt_not_mapped_to_genome_or_libs_stats_file),
                       name = "Average reads not \n mapped to genome or libs",
                       color = I("#2ca02c")) %>%
  plotly::layout(xaxis = xform) %>%
  plotly::layout(xaxis = list(title = "Sample"),
                 yaxis = list(title = "Read count"))
```

```{r, out.width = "100%"}
# create datatable to explore the RNA's that are found in the media
DT::datatable(mapping_rates_data %>%
                dplyr::mutate_if(is.numeric, format, big.mark = ",", scientific = FALSE, digits = 0) %>%
                dplyr::mutate(across(c("sample", "treatment", "low_fastq_read_count"), as.factor)) %>%
                dplyr::select(sample,
                              treatment,
                              raw_read_count_fastq_file,
                              excerpt_reads_used_for_alignment_stats_file,
                              excerpt_not_mapped_to_genome_or_libs_stats_file,
                              low_fastq_read_count),
              filter = "top",
              rownames = FALSE,
              colnames = c("Sample",
                           "Treatment",
                           "Raw sequencing read count",
                           "Reads used for alignment",
                           "Not mapped to genome or libs",
                           "Low raw sequencing read count"),
              extensions = list("ColReorder" = NULL,
                                "Buttons" = NULL,
                                "FixedColumns" = list(leftColumns=1)),
              options = list(
                dom = "BRrltpi",
                autoWidth = TRUE,
                columnDefs = list(list(width = "200px", targets = 0)),
                lengthMenu = list(c(10, 50, -1), c("10", "50", "All")),
                ColReorder = TRUE,
                buttons =
                  list("copy",
                       list(extend = "collection",
                            buttons = c("csv", "excel", "pdf"),
                            text = "Download"),
                       I("colvis"))))
```

## manual mapping

*Iterative mapping to the human reads of various databases in the order shown - each time, the unmapped reads from each step are mapped to the next database*

```{r, out.width = "100%"}
# read in mapping rates info and remove empty rows
manual_mapping_rates_data <- utils::read.csv("../manual_mapping/mapping_rates.csv") %>%
  tidyr::drop_na()

# double check the samples are in a consistant order - naturally sorted by sample names
manual_mapping_rates_data <- manual_mapping_rates_data[mixedorder(as.character(manual_mapping_rates_data$sample)),]

# flag samples with low read counts
manual_mapping_rates_data <- manual_mapping_rates_data %>%
  dplyr::mutate(low_fastq_read_count = dplyr::case_when(raw_read_count_fastq_file < config$low_sequencing_read_count ~ TRUE,
                                                        raw_read_count_fastq_file > config$low_sequencing_read_count ~ FALSE))

# join metadata with mapping rates data
manual_mapping_rates_data <- dplyr::left_join(manual_mapping_rates_data, metadata, by = "sample")

# create a bar chart of the raw read counts
manual_mapping_rates_data %>%
  plotly::plot_ly(x = ~sample,
                  y = ~raw_read_count_fastq_file,
                  type = "bar",
                  textposition = "none",
                  name = "Raw sequencing \n read count",
                  hoverinfo = "text",
                  text = ~base::paste("</br> Raw sequencing read count:", base::format(raw_read_count_fastq_file, big.mark = ",", scientific = FALSE, digits = 2),
                                      "</br> Sample:", sample,
                                      "</br> Treatment:", treatment)) %>%
  
  plotly::add_trace(y = ~miRBase_unmapped,
                    name = "miRBase unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> miRBase unmapped:", base::format(miRBase_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~PSICQUIC_unmapped,
                    name = "PSICQUIC unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> PSICQUIC unmapped:", base::format(PSICQUIC_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~TarBase_unmapped,
                    name = "TarBase unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> TarBase unmapped:", base::format(TarBase_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~piRNAdb_unmapped,
                    name = "piRNAdb unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> piRNAdb unmapped:", base::format(piRNAdb_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~piRBase_unmapped,
                    name = "piRBase unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> piRBase unmapped:", base::format(piRBase_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~SILVA_unmapped,
                    name = "SILVA unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> SILVA unmapped:", base::format(SILVA_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~Rfam_unmapped,
                    name = "Rfam unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Rfam unmapped:", base::format(Rfam_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~X5SrRNAdb_unmapped,
                    name = "5SrRNAdb unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> 5SrRNAdb unmapped:", base::format(X5SrRNAdb_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~Greengenes_unmapped,
                    name = "Greengenes unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Greengenes unmapped:", base::format(Greengenes_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~CircBank_unmapped,
                    name = "CircBank unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> CircBank unmapped:", base::format(CircBank_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~GtRNAdb_unmapped,
                    name = "GtRNAdb unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> GtRNAdb unmapped:", base::format(GtRNAdb_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~LncBook_unmapped,
                    name = "LncBook unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> LncBook unmapped:", base::format(LncBook_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~LNCipedia_unmapped,
                    name = "LNCipedia unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> LNCipedia unmapped:", base::format(LNCipedia_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~lncRNAdb_unmapped,
                    name = "lncRNAdb unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> lncRNAdb unmapped:", base::format(lncRNAdb_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~LncBase_unmapped,
                    name = "LncBase unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> LncBase unmapped:", base::format(LncBase_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~snoDB_unmapped,
                    name = "snoDB unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> snoDB unmapped:", base::format(snoDB_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~snOPY_unmapped,
                    name = "snOPY unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> snOPY unmapped:", base::format(snOPY_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~gencode_unmapped,
                    name = "gencode unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> gencode unmapped:", base::format(gencode_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~NONCODE_unmapped,
                    name = "NONCODE unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> NONCODE unmapped:", base::format(NONCODE_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~SRPDB_unmapped,
                    name = "SRPDB unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> SRPDB unmapped:", base::format(SRPDB_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~HGNC_unmapped,
                    name = "HGNC unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> HGNC unmapped:", base::format(HGNC_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~GeneCards_unmapped,
                    name = "GeneCards unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> GeneCards unmapped:", base::format(GeneCards_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~ENA_unmapped,
                    name = "ENA unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> ENA unmapped:", base::format(ENA_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  plotly::add_trace(y = ~Ensembl_unmapped,
                    name = "Ensembl unmapped",
                    hoverinfo = "text",
                    text = ~base::paste("</br> Ensembl unmapped:", base::format(Ensembl_unmapped, big.mark = ",", scientific = FALSE, digits = 2),
                                        "</br> Sample:", sample,
                                        "</br> Treatment:", treatment)) %>%
  # add line that show you the average raw read count across all samples
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(manual_mapping_rates_data$raw_read_count_fastq_file),
                       yend = base::mean(manual_mapping_rates_data$raw_read_count_fastq_file),
                       name = "Average raw sequencing \n read count",
                       color = I("#1f77b4")) %>%
  # add line that show you the average reads counts remaining after all mapping
  plotly::add_segments(x = dplyr::first(sample_list),
                       xend = dplyr::last(sample_list),
                       y = base::mean(manual_mapping_rates_data$Ensembl_unmapped),
                       yend = base::mean(manual_mapping_rates_data$Ensembl_unmapped),
                       name = "Average raw sequencing \n read count remaining after \n all mapping",
                       color = I("#1f77b4")) %>%
  plotly::layout(xaxis = xform) %>%
  plotly::layout(xaxis = list(title = "Sample"),
                 yaxis = list(title = "Read count"))
```

```{r, out.width = "100%"}
# create datatable to explore the RNA's that are found in the media
DT::datatable(manual_mapping_rates_data %>%
                dplyr::mutate_if(is.numeric, format, big.mark = ",", scientific = FALSE, digits = 0) %>%
                dplyr::mutate(across(c("sample", "treatment", "low_fastq_read_count"), as.factor)) %>%
                dplyr::select(sample,
                              treatment,
                              raw_read_count_fastq_file,
                              miRBase_unmapped,
                              PSICQUIC_unmapped,
                              TarBase_unmapped,
                              piRNAdb_unmapped,
                              piRBase_unmapped,
                              SILVA_unmapped,
                              Rfam_unmapped,
                              X5SrRNAdb_unmapped,
                              Greengenes_unmapped,
                              CircBank_unmapped,
                              GtRNAdb_unmapped,
                              LncBook_unmapped,
                              LNCipedia_unmapped,
                              lncRNAdb_unmapped,
                              LncBase_unmapped,
                              snoDB_unmapped,
                              snOPY_unmapped,
                              gencode_unmapped,
                              NONCODE_unmapped,
                              SRPDB_unmapped,
                              HGNC_unmapped,
                              GeneCards_unmapped,
                              ENA_unmapped,
                              Ensembl_unmapped,
                              low_fastq_read_count),
              filter = "top",
              rownames = FALSE,
              colnames = c("Sample",
                           "Treatment",
                           "Raw sequencing read count",
                           "miRBase",
                           "PSICQUIC",
                           "TarBase",
                           "piRNAdb",
                           "piRBase",
                           "SILVA",
                           "Rfam",
                           "5SrRNAdb",
                           "Greengenes",
                           "CircBank",
                           "GtRNAdb",
                           "LncBook",
                           "LNCipedia",
                           "lncRNAdb",
                           "LncBase",
                           "snoDB",
                           "snOPY",
                           "gencode",
                           "NONCODE",
                           "SRPDB",
                           "HGNC",
                           "GeneCards",
                           "ENA",
                           "Ensembl",
                           "Low raw sequencing read count"),
              extensions = list("ColReorder" = NULL,
                                "Buttons" = NULL,
                                "FixedColumns" = list(leftColumns=1)),
              options = list(
                dom = "BRrltpi",
                autoWidth = TRUE,
                columnDefs = list(list(width = "200px", targets = 0)),
                lengthMenu = list(c(10, 50, -1), c("10", "50", "All")),
                ColReorder = TRUE,
                buttons =
                  list("copy",
                       list(extend = "collection",
                            buttons = c("csv", "excel", "pdf"),
                            text = "Download"),
                       I("colvis"))))
```

```{r, cleanup, results = "hide"}
# clean up
rm(list = ls())
```
