---
title: "Differential expression analysis - limma/voom"
author:
  # - Jane Doe^[Institution Two, jane@example.org]      # add report authors (uncomment if using)
  # - John Doe^[Institution One, john@example.org]      # add a second report author (uncomment if using)
date: "Date: `r base::format(Sys.time(), '%d/%m/%Y')`"
always_allow_html: true
output:
  bookdown::html_document2:
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: true
    number_sections: false
    self_contained: yes
    css: ../esr-styles.css
    highlight: pygments
    includes:
      after_body: ../esr_footer.html
  editor_options: 
    chunk_output_type: console
---

<img src="../images/ESR_logo.svg" style="position:absolute;top:19px;right:40px;width:23%;" />

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, error = FALSE, message = FALSE, fig.align = "center")
```

```{bash}
# create a directory to write data to file (if it doesn't yet exist)
mkdir -p diff_expr_results/
```

```{r}
# load libraries
library(dplyr)
library(limma)
library(DESeq2)
library(janitor)
library(edgeR)
library(DT)
library(apeglm)
library(plotly)
library(heatmaply)
library(gtools)
library(textshape)
library(tidyr)

# for icons at bottom of doc
htmltools::tagList(rmarkdown::html_dependency_font_awesome())

# read in yaml config file
config <- yaml::yaml.load_file("../config/config.yaml")

# read in metadata
metadata <- utils::read.csv(base::file.path(config$metadata))
```

## Info

General info:

- Differential expression analysis using [limma/voom](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29)
- Some values have been rounded

Datasets analysed:

```{r, results = "asis"}
# print the datasets the user has chosen to analyse for this document
base::cat(base::paste0(" - mirna smrnaseq: ", config$mirna_smrnaseq, "\n",
                       " - mirna excerpt: ", config$mirna_excerpt, "\n",
                       " - pirna excerpt: ", config$pirna_excerpt, "\n",
                       " - trna excerpt: ", config$trna_excerpt, "\n",
                       " - circrna excerpt: ", config$circrna_excerpt, "\n",
                       " - gencode excerpt: ", config$gencode_excerpt, "\n"))
```

Treatment comparisons: 

```{r, results = "asis"}
# print the treatment group comparisons the user has chosen to analyse
base::cat(base::paste0("- ", config$contrasts, "\n"))
```

```{r, results = "asis"}
# print the number of samples analysed
base::cat(base::paste0("Total number of samples: ", base::length(base::unique(metadata$sample))))
```

Number of samples in each treatment group:

```{r, results = "asis"}
# print the number of samples in each treatment group
n_samples_by_treatment <- metadata %>%
  group_by(treatment) %>%
  dplyr::summarise(n_samples = n())

base::cat(base::paste0("- ", n_samples_by_treatment$treatment, ": ", n_samples_by_treatment$n_samples, "\n"))
```

```{r, results = "hide"}
# create a list defining which code chunks to analyse (that are set to TRUE for both the datasets and contrasts to analyse) based on the yaml user configuration file
to_eval_chunk <- config[c("mirna_smrnaseq",
                          "mirna_excerpt",
                          "pirna_excerpt",
                          "trna_excerpt",
                          "circrna_excerpt",
                          "gencode_excerpt")] %>%
  rep(times = 10)

names(to_eval_chunk)[1:6] <- paste(names(to_eval_chunk)[1:6], "1", sep = "_")
names(to_eval_chunk)[7:12] <- paste(names(to_eval_chunk)[7:12], "2", sep = "_")
names(to_eval_chunk)[13:18] <- paste(names(to_eval_chunk)[13:18], "3", sep = "_")
names(to_eval_chunk)[19:24] <- paste(names(to_eval_chunk)[19:24], "4", sep = "_")
names(to_eval_chunk)[25:30] <- paste(names(to_eval_chunk)[25:30], "5", sep = "_")
names(to_eval_chunk)[31:36] <- paste(names(to_eval_chunk)[31:36], "6", sep = "_")
names(to_eval_chunk)[37:42] <- paste(names(to_eval_chunk)[37:42], "7", sep = "_")
names(to_eval_chunk)[43:48] <- paste(names(to_eval_chunk)[43:48], "8", sep = "_")
names(to_eval_chunk)[49:54] <- paste(names(to_eval_chunk)[49:54], "9", sep = "_")
names(to_eval_chunk)[55:60] <- paste(names(to_eval_chunk)[55:60], "10", sep = "_")

n_contrasts <- base::length(config$contrasts)

to_eval_chunk <- if(n_contrasts == 1) {
  replace(to_eval_chunk, 7:60, FALSE)
} else if(n_contrasts == 2) {
  replace(to_eval_chunk, 13:60, FALSE)
} else if(n_contrasts == 3) {
  replace(to_eval_chunk, 19:60, FALSE)
} else if(n_contrasts == 4) {
  replace(to_eval_chunk, 25:60, FALSE)
} else if(n_contrasts == 5) {
  replace(to_eval_chunk, 31:60, FALSE)
} else if(n_contrasts == 6) {
  replace(to_eval_chunk, 37:60, FALSE)
} else if(n_contrasts == 7) {
  replace(to_eval_chunk, 43:60, FALSE)
} else if(n_contrasts == 8) {
  replace(to_eval_chunk, 49:60, FALSE)
} else if(n_contrasts == 9) {
  replace(to_eval_chunk, 55:60, FALSE)
} else if(n_contrasts == 10) {
  to_eval_chunk
}

# create a list defining which code chunks to analyse (based on the number of contrasts to analyse) based on the yaml user configuration file
contrasts_chunk_eval <- list(contrast_1 = TRUE,
                             contrast_2 = TRUE,
                             contrast_3 = TRUE,
                             contrast_4 = TRUE,
                             contrast_5 = TRUE,
                             contrast_6 = TRUE,
                             contrast_7 = TRUE,
                             contrast_8 = TRUE,
                             contrast_9 = TRUE,
                             contrast_10 = TRUE)

contrasts_chunk_eval <- if(n_contrasts == 1) {
  replace(contrasts_chunk_eval, 2:10, FALSE)
} else if(n_contrasts == 2) {
  replace(contrasts_chunk_eval, 3:10, FALSE)
} else if(n_contrasts == 3) {
  replace(contrasts_chunk_eval, 4:10, FALSE)
} else if(n_contrasts == 4) {
  replace(contrasts_chunk_eval, 5:10, FALSE)
} else if(n_contrasts == 5) {
  replace(contrasts_chunk_eval, 6:10, FALSE)
} else if(n_contrasts == 6) {
  replace(contrasts_chunk_eval, 7:10, FALSE)
} else if(n_contrasts == 7) {
  replace(contrasts_chunk_eval, 8:10, FALSE)
} else if(n_contrasts == 8) {
  replace(contrasts_chunk_eval, 9:10, FALSE)
} else if(n_contrasts == 9) {
  replace(contrasts_chunk_eval, 10, FALSE)
} else if(n_contrasts == 10) {
  contrasts_chunk_eval
}

# read in mapping rates info
mapping_rates <- utils::read.csv("../mapping_rates/mapping_rates.csv")

# tidy up sample names
mapping_rates$sample <- base::gsub('(_*)_\\w+', '', mapping_rates$sample)

# remove empty rows
mapping_rates <- mapping_rates %>%
  tidyr::drop_na()

# evalutate/setup minimum logFC threshold
min_logfc <- base::eval(base::parse(text = config$min_lfc))

# set up a flag for when the sequencing read counts for a sample is low (defined by the user in the config file)
low_read_count_flag <- mapping_rates %>%
  dplyr::filter(raw_read_count_fastq_file < config$low_sequencing_read_count) %>%
  dplyr::group_by(sample) %>%
  dplyr::summarise() %>%
  dplyr::pull(sample)

# make a TRUE/FALSE list that defines how many contrasts were analysed
# a vector of up to 6 elements since this is the max number of contrasts I've accounted for in this document
# will be used to conditionally include code chunks depending on how many contrasts/comparisons the user has chosen to analyse
contrasts <- base::append(config$contrasts, rep.int(FALSE, times = 10))
contrasts <- contrasts[1:10]
contrasts[contrasts != FALSE] <- TRUE
contrasts <- base::as.logical(contrasts)

# specify treatments by creating a string of conditions that match the order of the columns/samples in the count data
# get the treatments and samples names from the metadata file
treatments <- metadata %>%
  dplyr::select(sample, treatment)

# also sort by the sample column (important so it matches the order of the samples count datasets)
# this is critical for DESeq2 - it assumes they are in the same order
treatments <- treatments[gtools::mixedorder(base::as.character(treatments$sample)),]

# extract only the conditions/groups and create a list out of it
ordered_treatments <- treatments %>%
  dplyr::pull(treatment)
```

```{r, eval = config$mirna_smrnaseq}
# load count dataset (premade rds objects)
raw_mirna_smrnaseq_data <- base::readRDS(base::file.path(config$template_dir, "prepare_counts/rds_objects/raw_mirna_smrnaseq_counts.rds"))
```

```{r, eval = config$mirna_excerpt}
# load count dataset (premade rds objects)
raw_mirna_excerpt_data <- base::readRDS(base::file.path(config$template_dir, "prepare_counts/rds_objects/raw_mirna_excerpt_counts.rds"))
```

```{r, eval = config$pirna_excerpt}
# load count dataset (premade rds objects)
raw_pirna_excerpt_data <- base::readRDS(base::file.path(config$template_dir, "prepare_counts/rds_objects/raw_pirna_excerpt_counts.rds"))
```

```{r, eval = config$trna_excerpt}
# load count dataset (premade rds objects)
raw_trna_excerpt_data <- base::readRDS(base::file.path(config$template_dir, "prepare_counts/rds_objects/raw_trna_excerpt_counts.rds"))
```

```{r, eval = config$circrna_excerpt}
# load count dataset (premade rds objects)
raw_circrna_excerpt_data <- base::readRDS(base::file.path(config$template_dir, "prepare_counts/rds_objects/raw_circrna_excerpt_counts.rds"))
```

```{r, eval = config$gencode_excerpt}
# load count dataset (premade rds objects)
raw_gencode_excerpt_data <- base::readRDS(base::file.path(config$template_dir, "prepare_counts/rds_objects/raw_gencode_excerpt_counts.rds"))
```

```{r}
# make a list of all the count datasets in the global environment
count_datasets <- base::do.call("list", base::mget(base::grep("_data", base::names(.GlobalEnv), value=TRUE)))

# make sure the samples (columns) in all the datasets are in the same order for specifying the treatments downstream
# (that depends on the columns being in the correct order)  (loop over all count datasets)
count_datasets <- base::lapply(count_datasets, function(x) {
  
  x[ , gtools::mixedsort(names(x))]
  
})

# convert datasets to matrices (loop over all count datasets)
# also round read counts to the nearest integer to avoid a downstream error with DESeqDataSetFromMatrix() (see this discussion https://www.biostars.org/p/368158/)
count_datasets <- base::lapply(count_datasets, function(x) {
  
  x %>%
    base::round() %>%
    base::as.matrix(sep = "\t", row.names = "rna_id")
  
})

# setup a function to plot residual variances
# (code adapted from the "limma::plotSA()" function to make interactive and add more info like which RNA each point is)
plotSA_interactive <- function(data) {
  
  # get data from voom fit
  x <- data$Amean
  y <- base::sqrt(data$sigma)
  
  if (!(base::is.null(data$weights) || zero.weights)) {
    allzero <- base::rowSums(data$weights > 0, na.rm = TRUE) == 
      0
    y[allzero] <- NA
  }
  
  if (base::length(data$df.prior) > 1L) {
    df2 <- base::max(data$df.prior)
    s2 <- data$sigma^2/fit$s2.prior
    pdn <- stats::pf(s2, df1 = data$df.residual, df2 = df2)
    pup <- stats::pf(s2, df1 = data$df.residual, df2 = df2, lower.tail = FALSE)
    FDR <- stats::p.adjust(2 * pmin(pdn, pup), method = "BH")
    colv[FDR <= 0.5] <- col[2]
  }
  
  # make into a dataframe for plotting, also get the RNA info
  plotting_data <- base::cbind(x, y) %>%
    base::as.data.frame() %>%
    tibble::rownames_to_column("rna")
  
  # shorten the RNA names for plotting
  plotting_data$rna <- base::gsub("\\|.*", " etc.", plotting_data$rna)
  
  # calculate horizontal line
  h <- base::sqrt(base::sqrt(data$s2.prior))
  
  # set up function to draw horizontal line
  hline <- function(y = 0, color = "black") {
    base::list(
      type = "line", 
      x0 = 0, 
      x1 = 1, 
      xref = "paper",
      y0 = y, 
      y1 = y, 
      line = base::list(color = color)
    )
  }
  
  # plot
  plotting_data %>%
    plotly::plot_ly(x = ~x,
                    y = ~y,
                    marker = list(color = "#0097db", opacity = 0.7),
                    hoverinfo = "text",
                    text = ~ base::paste("</br> Average log expression:", base::format(x, big.mark = ",", scientific = FALSE, digits = 3),
                                         "</br> sqrt(sigma):", base::format(y, big.mark = ",", scientific = FALSE, digits = 4),
                                         "</br> RNA:", rna),
                    showlegend = FALSE) %>%
    plotly::layout(shapes = base::list(hline(h))) %>%
    plotly::layout(xaxis = base::list(title = "Average log expression"),
                   yaxis = base::list(title = "sqrt(sigma)"))
}
```

```{r, results = "asis"}
base::cat(base::paste0("The data was filtered using the [filterByExpr](https://rdrr.io/bioc/edgeR/man/filterByExpr.html) function. The filtering keeps RNA's that have count-per-million (CPM) above min.count (", config$min_count, ")", " in n samples (", base::length(base::unique(metadata$sample)), "). In addition, each kept RNA is required to have at least min.total.count reads (", config$min_total_count, ") across all the samples. From a statistical point of view, removing low count RNA's allows the mean-variance relationship in the data to be estimated with greater reliability ([Law et al., (2018)](https://f1000research.com/articles/5-1408))."))
```

```{r, results = "asis"}
base::cat(base::paste0("Normalization factors used to scale the raw library sizes are calculated using the [calcNormFactors](https://www.rdocumentation.org/packages/edgeR/versions/3.14.0/topics/calcNormFactors) function using the TMM normalization method. This method uses the weighted trimmed mean of M-values (to the reference) proposed by [Robinson and Oshlack (2010)](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25), where the weights are from the delta method on Binomial data. "))
```

```{r, results = "asis"}
base::cat(base::paste0("The count data is then transformed to log2-counts per million (logCPM) using the [voom](https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/voom) function (see the paper [here](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29)). Fitting linear models to the comparisons/contrasts (", base::toString(config$contrasts), ") is carried out using the [lmFit](https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/lmFit) and [contrasts.fit](https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/contrasts.fit) functions. Next, empirical Bayes moderation is carried using the [eBayes](https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/ebayes) function which borrows information across all RNA's to obtain more precise estimates of RNA-wise variability ([Law et al., (2018)](https://f1000research.com/articles/5-1408)). It computes moderated t-statistics, moderated F-statistic, and log-odds of differential expression by empirical Bayes moderation of the standard errors towards a common value."))
```

```{r, results = "asis"}
base::cat(base::paste0("The Benjamini and Hochberg method (see [Benjamini & Hochberg (1995)](https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.2517-6161.1995.tb02031.x)) was used in the [topTreat function](https://www.rdocumentation.org/packages/limma/versions/3.28.14/topics/toptable) to adjust the p-values to account for multiple testing, reducing the false discovery rate. See [this article comparing the different adjustment methods](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6099145/). Both the limma/voom and DESeq2 differential expression methods use the same p-value adjustment method."))
```

```{r, limma_voom}
# create DGEList objects for the datasets (loop over all count datasets)
dge <- base::lapply(count_datasets, function(x) {
  
  edgeR::DGEList(counts = x, group = ordered_treatments)
  
})

# filter lowly expressed RNA
keep <- base::lapply(dge, function(x) {
  
  edgeR::filterByExpr(x, group=ordered_treatments, min.count = config$min_count, min.total.count = config$min_total_count)
  
})
```

```{r, eval = config$mirna_smrnaseq}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_mirna_smrnaseq_filtered <- dge$raw_mirna_smrnaseq_data[keep$raw_mirna_smrnaseq_data,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r, eval = config$mirna_excerpt}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_mirna_excerpt_filtered <- dge$raw_mirna_excerpt_data[keep$raw_mirna_excerpt_data,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r, eval = config$pirna_excerpt}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_pirna_excerpt_filtered <- dge$raw_pirna_excerpt_data[keep$raw_pirna_excerpt_data,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r, eval = config$trna_excerpt}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_trna_excerpt_filtered <- dge$raw_trna_excerpt_data[keep$raw_trna_excerpt_data,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r, eval = config$circrna_excerpt}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_circrna_excerpt_filtered <- dge$raw_circrna_excerpt_data[keep$raw_circrna_excerpt_data,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r, eval = config$gencode_excerpt}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_gencode_excerpt_filtered <- dge$raw_gencode_excerpt_data[keep$raw_gencode_excerpt_data,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r}
# create a model matrix
# also simplify the column names in the model matrix (loop over all count datasets)
model_matrix <- base::lapply(dge, function(x) {
  
  stats::model.matrix(~0+ordered_treatments) %>%
    base::as.data.frame() %>%
    dplyr::rename_with(~ (base::gsub("ordered_treatments", "", .x, fixed = TRUE))) %>%
    base::data.matrix()
  
})

# specify which groups to compare (based on the users choice of contrasts to compare) (loop over all count datasets)
contr_matrix <- base::lapply(model_matrix, function(x) {
  
  limma::makeContrasts(contrasts = c(config$contrasts), levels = base::colnames(x))
  
})
```

## Mean variance plot {.tabset .tabset-fade}

The mean-variance relationship of log-CPM values for this dataset. Typically, the “voom-plot” shows a decreasing trend between the means and variances resulting from a combination of technical variation in the sequencing experiment and biological variation amongst the replicate samples from different treatment groups. Experiments with high biological variation usually result in flatter trends, where variance values plateau at high expression values. Experiments with low biological variation tend to result in sharp decreasing trends ([Law et al., (2018)](https://f1000research.com/articles/5-1408)).

Moreover, the voom-plot provides a visual check on the level of filtering performed upstream. If filtering of lowly-expressed RNA's is insufficient, a drop in variance levels can be observed at the low end of the expression scale due to very small counts. If this is observed, one should return to the earlier filtering step and increase the expression threshold applied to the dataset ([Law et al., (2018)](https://f1000research.com/articles/5-1408)).

```{r, eval = config$mirna_smrnaseq, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### miRNA's (smrnaseq)")
```

```{r, eval = config$mirna_smrnaseq, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
mirna_smrnaseq_limmavoom_v <- limma::voom(dge_mirna_smrnaseq_filtered,  model_matrix$raw_mirna_smrnaseq_data, plot=TRUE)
```

```{r, eval = config$mirna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### miRNA's (excerpt)")
```

```{r, eval = config$mirna_excerpt, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
mirna_excerpt_limmavoom_v <- limma::voom(dge_mirna_excerpt_filtered,  model_matrix$raw_mirna_excerpt_data, plot=TRUE)
```

```{r, eval = config$pirna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### piRNA's (excerpt)")
```

```{r, eval = config$pirna_excerpt, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
pirna_excerpt_limmavoom_v <- limma::voom(dge_pirna_excerpt_filtered,  model_matrix$raw_pirna_excerpt_data, plot=TRUE)
```

```{r, eval = config$trna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### tRNA's (excerpt)")
```

```{r, eval = config$trna_excerpt, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
trna_excerpt_limmavoom_v <- limma::voom(dge_trna_excerpt_filtered,  model_matrix$raw_trna_excerpt_data, plot=TRUE)
```

```{r, eval = config$circrna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### circRNA's (excerpt)")
```

```{r, eval = config$circrna_excerpt, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
circrna_excerpt_limmavoom_v <- limma::voom(dge_circrna_excerpt_filtered,  model_matrix$raw_circrna_excerpt_data, plot=TRUE)
```

```{r, eval = config$gencode_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### gencodes (excerpt)")
```

```{r, eval = config$gencode_excerpt, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
gencode_excerpt_limmavoom_v <- limma::voom(dge_gencode_excerpt_filtered,  model_matrix$raw_gencode_excerpt_data, plot=TRUE)
```

## Residual variances {.tabset .tabset-fade}

An adapted version of the [plotSA](http://rstudio.esr.cri.nz:8787/s/dec5c0fe7569e683b7632/help/library/limma/html/plotSA.html) function was used to create this plot (made interactive and included individual RNA information for each data point), which can be used to check the mean-variance relationship of the expression data, after fitting a linear model.

```{r, eval = config$mirna_smrnaseq, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### miRNA's (smrnaseq)")
```

```{r, eval = config$mirna_smrnaseq, out.width = "100%"}
# fit model to the expression values for each RNA
mirna_smrnaseq_limmavoom_vfit <- limma::lmFit(mirna_smrnaseq_limmavoom_v, model_matrix$raw_mirna_smrnaseq_data)

# estimate contrast for each miRNA
mirna_smrnaseq_limmavoom_vfit <- limma::contrasts.fit(mirna_smrnaseq_limmavoom_vfit, contr_matrix$raw_mirna_smrnaseq_data)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
mirna_smrnaseq_limmavoom_efit <- limma::eBayes(mirna_smrnaseq_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(mirna_smrnaseq_limmavoom_efit)
```

```{r, eval = config$mirna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### miRNA's (excerpt)")
```

```{r, eval = config$mirna_excerpt, out.width = "100%"}
# fit model to the expression values for each RNA
mirna_excerpt_limmavoom_vfit <- limma::lmFit(mirna_excerpt_limmavoom_v, model_matrix$raw_mirna_excerpt_data)

# estimate contrast for each miRNA
mirna_excerpt_limmavoom_vfit <- limma::contrasts.fit(mirna_excerpt_limmavoom_vfit, contr_matrix$raw_mirna_excerpt_data)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
mirna_excerpt_limmavoom_efit <- limma::eBayes(mirna_excerpt_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(mirna_excerpt_limmavoom_efit)
```

```{r, eval = config$pirna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### piRNA's (excerpt)")
```

```{r, eval = config$pirna_excerpt, out.width = "100%"}
# fit model to the expression values for each RNA
pirna_excerpt_limmavoom_vfit <- limma::lmFit(pirna_excerpt_limmavoom_v, model_matrix$raw_pirna_excerpt_data)

# estimate contrast for each miRNA
pirna_excerpt_limmavoom_vfit <- limma::contrasts.fit(pirna_excerpt_limmavoom_vfit, contr_matrix$raw_pirna_excerpt_data)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
pirna_excerpt_limmavoom_efit <- limma::eBayes(pirna_excerpt_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(pirna_excerpt_limmavoom_efit)
```

```{r, eval = config$trna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### tRNA's (excerpt)")
```

```{r, eval = config$trna_excerpt, out.width = "100%"}
# fit model to the expression values for each RNA
trna_excerpt_limmavoom_vfit <- limma::lmFit(trna_excerpt_limmavoom_v, model_matrix$raw_trna_excerpt_data)

# estimate contrast for each miRNA
trna_excerpt_limmavoom_vfit <- limma::contrasts.fit(trna_excerpt_limmavoom_vfit, contr_matrix$raw_trna_excerpt_data)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
trna_excerpt_limmavoom_efit <- limma::eBayes(trna_excerpt_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(trna_excerpt_limmavoom_efit)
```

```{r, eval = config$circrna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### circRNA's (excerpt)")
```

```{r, eval = config$circrna_excerpt, out.width = "100%"}
# fit model to the expression values for each RNA
circrna_excerpt_limmavoom_vfit <- limma::lmFit(circrna_excerpt_limmavoom_v, model_matrix$raw_circrna_excerpt_data)

# estimate contrast for each miRNA
circrna_excerpt_limmavoom_vfit <- limma::contrasts.fit(circrna_excerpt_limmavoom_vfit, contr_matrix$raw_circrna_excerpt_data)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
circrna_excerpt_limmavoom_efit <- limma::eBayes(circrna_excerpt_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(circrna_excerpt_limmavoom_efit)
```

```{r, eval = config$gencode_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("### gencodes (excerpt)")
```

```{r, eval = config$gencode_excerpt, out.width = "100%"}
# fit model to the expression values for each RNA
gencode_excerpt_limmavoom_vfit <- limma::lmFit(gencode_excerpt_limmavoom_v, model_matrix$raw_gencode_excerpt_data)

# estimate contrast for each miRNA
gencode_excerpt_limmavoom_vfit <- limma::contrasts.fit(gencode_excerpt_limmavoom_vfit, contr_matrix$raw_gencode_excerpt_data)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
gencode_excerpt_limmavoom_efit <- limma::eBayes(gencode_excerpt_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(gencode_excerpt_limmavoom_efit)
```

```{r}
# make a list of all the efit objects in the global environment
efit <- base::do.call("list", base::mget(base::grep("_efit", base::names(.GlobalEnv), value=TRUE)))
```

```{r, eval = contrasts[1]}
# extract the results of the limma/voom differential expression analysis (first contrast) (loop over all efit datasets)
limma_voom_results_1 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 1,
                    n = Inf,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_1 <- base::do.call(rbind, limma_voom_results_1)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_1 <- limma_voom_results_1 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("(_excerpt)|(_smrnaseq)", "", source)) %>%
  dplyr::mutate(pipeline = gsub("(mirna_)|(pirna_)|(trna_)|(circrna_)|(gencode_)", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_1 <- limma_voom_results_1 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_1 <- limma_voom_results_1 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)

# write differential expression results to file
limma_voom_results_1 %>%
  utils::write.table(file = "./diff_expr_results/limma_voom_diff_expr_results_comparison_1.tsv", row.names=FALSE, sep="\t")
```

```{r, eval = contrasts[2]}
# extract the results of the limma/voom differential expression analysis (second contrast) (loop over all efit datasets)
limma_voom_results_2 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 2,
                    n = Inf,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_2 <- base::do.call(rbind, limma_voom_results_2)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_2 <- limma_voom_results_2 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("(_excerpt)|(_smrnaseq)", "", source)) %>%
  dplyr::mutate(pipeline = gsub("(mirna_)|(pirna_)|(trna_)|(circrna_)|(gencode_)", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[2]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_2 <- limma_voom_results_2 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_2 <- limma_voom_results_2 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)

# write differential expression results to file
limma_voom_results_2 %>%
  utils::write.table(file = "./diff_expr_results/limma_voom_diff_expr_results_comparison_2.tsv", row.names=FALSE, sep="\t")
```

```{r, eval = contrasts[3]}
# extract the results of the limma/voom differential expression analysis (third contrast) (loop over all efit datasets)
limma_voom_results_3 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 3,
                    n = Inf,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_3 <- base::do.call(rbind, limma_voom_results_3)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_3 <- limma_voom_results_3 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("(_excerpt)|(_smrnaseq)", "", source)) %>%
  dplyr::mutate(pipeline = gsub("(mirna_)|(pirna_)|(trna_)|(circrna_)|(gencode_)", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[3]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_3 <- limma_voom_results_3 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_3 <- limma_voom_results_3 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)

# write differential expression results to file
limma_voom_results_3 %>%
  utils::write.table(file = "./diff_expr_results/limma_voom_diff_expr_results_comparison_3.tsv", row.names=FALSE, sep="\t")
```

```{r, eval = contrasts[4]}
# extract the results of the limma/voom differential expression analysis (fourth contrast) (loop over all efit datasets)
limma_voom_results_4 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 4,
                    n = Inf,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_4 <- base::do.call(rbind, limma_voom_results_4)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_4 <- limma_voom_results_4 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("(_excerpt)|(_smrnaseq)", "", source)) %>%
  dplyr::mutate(pipeline = gsub("(mirna_)|(pirna_)|(trna_)|(circrna_)|(gencode_)", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[4]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_4 <- limma_voom_results_4 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_4 <- limma_voom_results_4 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)

# write differential expression results to file
limma_voom_results_4 %>%
  utils::write.table(file = "./diff_expr_results/limma_voom_diff_expr_results_comparison_4.tsv", row.names=FALSE, sep="\t")
```

```{r, eval = contrasts[5]}
# extract the results of the limma/voom differential expression analysis (fifth contrast) (loop over all efit datasets)
limma_voom_results_5 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 5,
                    n = Inf,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_5 <- base::do.call(rbind, limma_voom_results_5)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_5 <- limma_voom_results_5 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("(_excerpt)|(_smrnaseq)", "", source)) %>%
  dplyr::mutate(pipeline = gsub("(mirna_)|(pirna_)|(trna_)|(circrna_)|(gencode_)", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[5]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_5 <- limma_voom_results_5 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_5 <- limma_voom_results_5 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)

# write differential expression results to file
limma_voom_results_5 %>%
  utils::write.table(file = "./diff_expr_results/limma_voom_diff_expr_results_comparison_5.tsv", row.names=FALSE, sep="\t")
```

```{r, eval = contrasts[6]}
# extract the results of the limma/voom differential expression analysis (sixth contrast) (loop over all efit datasets)
limma_voom_results_6 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 6,
                    n = Inf,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_6 <- base::do.call(rbind, limma_voom_results_6)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_6 <- limma_voom_results_6 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("(_excerpt)|(_smrnaseq)", "", source)) %>%
  dplyr::mutate(pipeline = gsub("(mirna_)|(pirna_)|(trna_)|(circrna_)|(gencode_)", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[6]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_6 <- limma_voom_results_6 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_6 <- limma_voom_results_6 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)

# write differential expression results to file
limma_voom_results_6 %>%
  utils::write.table(file = "./diff_expr_results/limma_voom_diff_expr_results_comparison_6.tsv", row.names=FALSE, sep="\t")
```

```{r, eval = contrasts[7]}
# extract the results of the limma/voom differential expression analysis (sixth contrast) (loop over all efit datasets)
limma_voom_results_7 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 7,
                    n = Inf,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_7 <- base::do.call(rbind, limma_voom_results_7)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_7 <- limma_voom_results_7 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("(_excerpt)|(_smrnaseq)", "", source)) %>%
  dplyr::mutate(pipeline = gsub("(mirna_)|(pirna_)|(trna_)|(circrna_)|(gencode_)", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[7]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_7 <- limma_voom_results_7 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_7 <- limma_voom_results_7 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)

# write differential expression results to file
limma_voom_results_7 %>%
  utils::write.table(file = "./diff_expr_results/limma_voom_diff_expr_results_comparison_7.tsv", row.names=FALSE, sep="\t")
```

```{r, eval = contrasts[8]}
# extract the results of the limma/voom differential expression analysis (sixth contrast) (loop over all efit datasets)
limma_voom_results_8 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 8,
                    n = Inf,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_8 <- base::do.call(rbind, limma_voom_results_8)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_8 <- limma_voom_results_8 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("(_excerpt)|(_smrnaseq)", "", source)) %>%
  dplyr::mutate(pipeline = gsub("(mirna_)|(pirna_)|(trna_)|(circrna_)|(gencode_)", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[8]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_8 <- limma_voom_results_8 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_8 <- limma_voom_results_8 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)

# write differential expression results to file
limma_voom_results_8 %>%
  utils::write.table(file = "./diff_expr_results/limma_voom_diff_expr_results_comparison_8.tsv", row.names=FALSE, sep="\t")
```

```{r, eval = contrasts[9]}
# extract the results of the limma/voom differential expression analysis (sixth contrast) (loop over all efit datasets)
limma_voom_results_9 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 9,
                    n = Inf,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_9 <- base::do.call(rbind, limma_voom_results_9)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_9 <- limma_voom_results_9 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("(_excerpt)|(_smrnaseq)", "", source)) %>%
  dplyr::mutate(pipeline = gsub("(mirna_)|(pirna_)|(trna_)|(circrna_)|(gencode_)", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[9]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_9 <- limma_voom_results_9 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_9 <- limma_voom_results_9 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)

# write differential expression results to file
limma_voom_results_9 %>%
  utils::write.table(file = "./diff_expr_results/limma_voom_diff_expr_results_comparison_9.tsv", row.names=FALSE, sep="\t")
```

```{r, eval = contrasts[10]}
# extract the results of the limma/voom differential expression analysis (sixth contrast) (loop over all efit datasets)
limma_voom_results_10 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 10,
                    n = Inf,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_10 <- base::do.call(rbind, limma_voom_results_10)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_10 <- limma_voom_results_10 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("(_excerpt)|(_smrnaseq)", "", source)) %>%
  dplyr::mutate(pipeline = gsub("(mirna_)|(pirna_)|(trna_)|(circrna_)|(gencode_)", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[10]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_10 <- limma_voom_results_10 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_10 <- limma_voom_results_10 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)

# write differential expression results to file
limma_voom_results_10 %>%
  utils::write.table(file = "./diff_expr_results/limma_voom_diff_expr_results_comparison_10.tsv", row.names=FALSE, sep="\t")
```

```{r}
# make a list of all the differential expression results objects in the global environment
limma_voom_diff_expr_results <- base::do.call("list", base::mget(base::grep("limma_voom_results_*", base::names(.GlobalEnv), value=TRUE)))

# collapse all the differential expression resultsresults from all the contrasts analysed
limma_voom_diff_expr_results  <- base::do.call(rbind, limma_voom_diff_expr_results)

# create a column that defines if an RNA is significant or not (at three significance levels and based on
# both limma/voom and deseq results). I ordered the case_when conditions from not significant to significant
# at the 1% level. This way, if an observation is true in several of these case_when categories, it'll be 
# marked with the smallest p-value category is fits into, sort of "rounded up" to the smallest p-value,
# because this case_when function seems to overwrite a value if it falls in a later case_when category
# This data will be used as a "flag" (using highlighting) in the downstream app to indicate a possibly
# significantly differentiated RNA the user of the app might want to investigate
limma_voom_diff_expr_results <- limma_voom_diff_expr_results %>%
  dplyr::mutate(significance = dplyr::case_when(adj_p_value > 0.10 ~ "greater_than_10%",
                                                (adj_p_value > 0.05) & (adj_p_value <= 0.10) ~ "significant_10%",
                                                (adj_p_value > 0.01) & (adj_p_value <= 0.05) ~ "significant_5%",
                                                (adj_p_value <= 0.01) ~ "significant_1%"))
```

```{r}
# write differential expression results to file
limma_voom_diff_expr_results %>%
  utils::write.table(file = "./diff_expr_results/all_limma_voom_diff_expr_results.tsv", row.names=FALSE, sep="\t")
```

```{r, cleanup, results = "hide"}
# clean up
rm(list = ls())
```
