---
title: "Differential expression analysis"
author:
  # - Jane Doe^[Institution Two, jane@example.org]      # add report authors (uncomment if using)
  # - John Doe^[Institution One, john@example.org]      # add a second report author (uncomment if using)
date: "Date: `r base::format(Sys.time(), '%d/%m/%Y')`"
output:
  html_document:
    code_download: true
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: true
  editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, error = FALSE, message = FALSE, fig.align = "center")
```

```{bash}
# create a directory to write data to file (if it doesn't yet exist)
mkdir -p diff_expr_results/
```

## Info

General info:

- Two methods were employed to undertake a differential expression analysis, namely [limma/voom](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29) and [deseq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8).

```{r, data_prep, results = "hide"}
# load libraries
library(dplyr)
library(limma)
library(DESeq2)
library(janitor)
library(edgeR)
library(DT)
library(apeglm)
library(plotly)
library(heatmaply)
library(gtools)
library(textshape)
library(tidyr)

# read in yaml config file
config <- yaml::yaml.load_file("../config/config.yaml")

# read in metadata
metadata <- utils::read.csv(base::file.path(config$metadata))

# load all the count datasets
mirna_excerpt_data <- utils::read.table(base::file.path(config$excerpt_merged_results_dir,
                                                        "exceRpt_miRNA_ReadCounts.txt"),
                                        header = TRUE,
                                        stringsAsFactors = FALSE,
                                        check.names = FALSE) %>%
  # remove S*_R1.fastq suffix from the sample/column names
  dplyr::rename_with(~ base::sub("[_][S]\\d+[_]R1.fastq", "", .))

pirna_excerpt_data <- utils::read.table(base::file.path(config$excerpt_merged_results_dir,
                                                        "exceRpt_piRNA_ReadCounts.txt"),
                                        header = TRUE,
                                        stringsAsFactors = FALSE,
                                        check.names = FALSE) %>%
  # remove S*_R1.fastq suffix from the sample/column names
  dplyr::rename_with(~ base::sub("[_][S]\\d+[_]R1.fastq", "", .))

trna_excerpt_data <- utils::read.table(base::file.path(config$excerpt_merged_results_dir,
                                                       "exceRpt_tRNA_ReadCounts.txt"),
                                       header = TRUE,
                                       stringsAsFactors = FALSE,
                                       check.names = FALSE) %>%
  # remove S*_R1.fastq suffix from the sample/column names
  dplyr::rename_with(~ base::sub("[_][S]\\d+[_]R1.fastq", "", .))

circrna_excerpt_data <- utils::read.table(base::file.path(config$excerpt_merged_results_dir,
                                                          "exceRpt_circularRNA_ReadCounts.txt"),
                                          header = TRUE,
                                          stringsAsFactors = FALSE,
                                          comment.char = "") %>%
  # remove S*_R1.fastq suffix from the sample/column names
  dplyr::rename_with(~ base::sub("[_][S]\\d+[_]R1.fastq", "", .))

gencode_excerpt_data <- utils::read.table(base::file.path(config$excerpt_merged_results_dir,
                                                          "exceRpt_gencode_ReadCounts.txt"),
                                          header = TRUE,
                                          stringsAsFactors = FALSE,
                                          check.names = FALSE) %>%
  # remove S*_R1.fastq suffix from the sample/column names
  dplyr::rename_with(~ base::sub("[_][S]\\d+[_]R1.fastq", "", .))

mirna_smrnaseq_data <- utils::read.table(base::file.path(config$smrnaseq_results_dir,
                                                         "/mirtop/mirtop.tsv"),
                                                         header = TRUE) %>%
  # get only the count data
  dplyr::select(-c("UID", "Read", "Variant", "iso_5p", "iso_3p", "iso_add3p", "iso_snp", "iso_5p_nt", "iso_3p_nt", "iso_add3p_nt", "iso_snp_nt")) %>%
  # "squash" out the isoform info (of the mirtop from smrnaseq pipeline)
  dplyr::group_by(miRNA) %>%
  dplyr::summarise_each(sum) %>%
  # convert column to rowname so the data can be normalised later
  tibble::column_to_rownames(var="miRNA")

# formatting to make counts datasets consistent between smrnaseq and excerpt pipelines
mirna_smrnaseq_data <- mirna_smrnaseq_data %>%
  # make all count data numeric
  dplyr::mutate_all(as.numeric)

  # replace "." with "-" so RNA names are the same
base::row.names(mirna_smrnaseq_data) <- base::gsub("\\.", "\\-", base::row.names(mirna_smrnaseq_data))

# create a list defining which code chunks to analyse (that are set to TRUE for both the datasets and contrasts to analyse) based on the yaml user configuration file
to_eval_chunk <- config[c("mirna_smrnaseq",
                          "mirna_excerpt",
                          "pirna_excerpt",
                          "trna_excerpt",
                          "circrna_excerpt",
                          "gencode_excerpt")] %>%
  rep(times = 6)

names(to_eval_chunk)[1:6] <- paste(names(to_eval_chunk)[1:6], "1", sep = "_")
names(to_eval_chunk)[7:12] <- paste(names(to_eval_chunk)[7:12], "2", sep = "_")
names(to_eval_chunk)[13:18] <- paste(names(to_eval_chunk)[13:18], "3", sep = "_")
names(to_eval_chunk)[19:24] <- paste(names(to_eval_chunk)[19:24], "4", sep = "_")
names(to_eval_chunk)[25:30] <- paste(names(to_eval_chunk)[25:30], "5", sep = "_")
names(to_eval_chunk)[31:36] <- paste(names(to_eval_chunk)[31:36], "6", sep = "_")

n_contrasts <- length(config$contrasts)

to_eval_chunk <- if(n_contrasts == 1) {
  replace(to_eval_chunk, 7:36, FALSE)
} else if(n_contrasts == 2) {
  replace(to_eval_chunk, 13:36, FALSE)
} else if(n_contrasts == 3) {
  replace(to_eval_chunk, 19:36, FALSE)
} else if(n_contrasts == 4) {
  replace(to_eval_chunk, 25:36, FALSE)
} else if(n_contrasts == 5) {
  replace(to_eval_chunk, 31:36, FALSE)
} else if(n_contrasts == 6) {
  to_eval_chunk
}

# create a list defining which code chunks to analyse (based on the number of contrasts to analyse) based on the yaml user configuration file
contrasts_chunk_eval <- list(contrast_1 = TRUE,
                             contrast_2 = TRUE,
                             contrast_3 = TRUE,
                             contrast_4 = TRUE,
                             contrast_5 = TRUE,
                             contrast_6 = TRUE)

contrasts_chunk_eval <- if(n_contrasts == 1) {
  replace(contrasts_chunk_eval, 2:6, FALSE)
} else if(n_contrasts == 2) {
  replace(contrasts_chunk_eval, 3:6, FALSE)
} else if(n_contrasts == 3) {
  replace(contrasts_chunk_eval, 4:6, FALSE)
} else if(n_contrasts == 4) {
  replace(contrasts_chunk_eval, 5:6, FALSE)
} else if(n_contrasts == 5) {
  replace(contrasts_chunk_eval, 6, FALSE)
} else if(n_contrasts == 6) {
  contrasts_chunk_eval
}

# create a vector defining the count datasets to analyse (that are set to TRUE) based on the yaml user configuration file
to_analyse <- config[c("mirna_smrnaseq",
                       "mirna_excerpt",
                       "pirna_excerpt",
                       "trna_excerpt",
                       "circrna_excerpt",
                       "gencode_excerpt")] %>%
  base::as.data.frame() %>%
  base::t() %>%
  base::as.data.frame() %>%
  tibble::rownames_to_column("rna_species") %>%
  dplyr::rename("analyse" = "V1") %>%
  tidyr::separate(rna_species, c("rna_species", "pipeline")) %>%
  dplyr::filter(analyse == TRUE) %>%
  dplyr::mutate(dataset_name = base::paste0(rna_species, "_", pipeline)) %>%
  dplyr::pull(dataset_name)

# set up a vector of all the possible datasets to analyse
count_datasets <- base::list(mirna_smrnaseq = mirna_smrnaseq_data,
                             mirna_excerpt = mirna_excerpt_data,
                             pirna_excerpt = pirna_excerpt_data,
                             trna_excerpt = trna_excerpt_data,
                             circrna_excerpt = circrna_excerpt_data,
                             gencode_excerpt = gencode_excerpt_data)

# filter all possible datasets by the datasets the user has specified to analyse
count_datasets <- count_datasets[to_analyse]

# read in mapping rates info
mapping_rates_data <- utils::read.csv("../mapping_rates/mapping_rates.csv")

# tidy up sample names
mapping_rates_data$sample <- base::gsub('(_*)_\\w+', '', mapping_rates_data$sample)

# remove empty rows
mapping_rates_data <- mapping_rates_data %>%
  tidyr::drop_na()

# evalutate/setup minimum logFC threshold
min_logfc <- base::eval(base::parse(text = config$min_lfc))

# set up a flag for when the sequencing read counts for a sample is low (defined by the user in the config file)
low_read_count_flag <- mapping_rates_data %>%
  dplyr::filter(raw_read_count_fastq_file < config$low_sequencing_read_count) %>%
  dplyr::group_by(sample) %>%
  dplyr::summarise() %>%
  dplyr::pull(sample)

# make a TRUE/FALSE list that defines how many contrasts were analysed
# a vector of up to 6 elements since this is the max number of contrasts I've accounted for in this document
# will be used to conditionally include code chunks depending on how many contrasts/comparsions the user has chosen to analyse
contrasts <- base::append(config$contrasts, rep.int(FALSE, times = 6))
contrasts <- contrasts[1:6]
contrasts[contrasts != FALSE] <- TRUE
contrasts <- base::as.logical(contrasts)

# specify treatments by creating a string of conditions that match the order of the columns/samples in the count data
# get the treatments and samples names from the metadata file
treatments <- metadata %>%
  dplyr::select(sample, treatment)

# also sort by the sample column (important so it matches the order of the samples count datasets)
# this is critical for DESeq2 - it assumes they are in the same order
treatments$sample <- gtools::mixedsort(treatments$sample)

# extract only the conditions/groups and create a list out of it
ordered_treatments <- treatments %>%
  dplyr::pull(treatment)

# make sure the samples (columns) in all the datasets are in the same order for specifying the treatments downstream
# (that depends on the columns being in the correct order)  (loop over all count datasets)
count_datasets<- base::lapply(count_datasets, function(x) {
  
  x[ , gtools::mixedsort(names(x))]
  
})

# convert datasets to matrices (loop over all count datasets)
# also round read counts to the nearest integer to avoid a downstream error with DESeqDataSetFromMatrix() (see this discussion https://www.biostars.org/p/368158/)
count_datasets <- base::lapply(count_datasets, function(x) {
  
  x %>%
    base::round() %>%
    base::as.matrix(sep = "\t", row.names = "rna_id")
  
})

# setup a function to plot residual variances
# (code adapted from the "limma::plotSA()" function to make interactive and add more info like which RNA each point is)
plotSA_interactive <- function(data) {
  
  # get data from voom fit
  x <- data$Amean
  y <- base::sqrt(data$sigma)
  
  if (!(base::is.null(data$weights) || zero.weights)) {
    allzero <- base::rowSums(data$weights > 0, na.rm = TRUE) == 
      0
    y[allzero] <- NA
  }
  
  if (length(data$df.prior) > 1L) {
    df2 <- base::max(data$df.prior)
    s2 <- data$sigma^2/fit$s2.prior
    pdn <- stats::pf(s2, df1 = data$df.residual, df2 = df2)
    pup <- stats::pf(s2, df1 = data$df.residual, df2 = df2, lower.tail = FALSE)
    FDR <- stats::p.adjust(2 * pmin(pdn, pup), method = "BH")
    colv[FDR <= 0.5] <- col[2]
  }
  
  # make into a dataframe for plotting, also get the RNA info
  plotting_data <- base::cbind(x, y) %>%
    base::as.data.frame() %>%
    tibble::rownames_to_column("rna")
  
  # shorten the RNA names for plotting
  plotting_data$rna <- base::gsub("\\|.*", " etc.", plotting_data$rna)
  
  # calculate horizontal line
  h <- base::sqrt(base::sqrt(data$s2.prior))
  
  # set up function to draw horizontal line
  hline <- function(y = 0, color = "blue") {
    base::list(
      type = "line", 
      x0 = 0, 
      x1 = 1, 
      xref = "paper",
      y0 = y, 
      y1 = y, 
      line = base::list(color = color)
    )
  }
  
  # plot
  plotting_data %>%
    plotly::plot_ly(x = ~x,
                    y = ~y,
                    hoverinfo = "text",
                    text = ~ base::paste("</br> Average log expression:", base::format(x, big.mark = ",", scientific = FALSE, digits = 2),
                                         "</br> sqrt(sigma):", base::format(y, big.mark = ",", scientific = FALSE, digits = 2),
                                         "</br> RNA:", rna)) %>%
    plotly::layout(shapes = base::list(hline(h))) %>%
    plotly::layout(xaxis = base::list(title = "Average log expression"),
                   yaxis = base::list(title = "sqrt(sigma)"))
}
```

Treatment comparisons: 

```{r, results = "asis"}
# print the treatment group comparisons the user has chosen to analyse
base::cat(base::paste0("- ", config$contrasts, "\n"))
```

```{r, results = "asis"}
# print the number of samples analysed
base::cat(base::paste0("Total number of samples: ", length(unique(metadata$sample))))
```

Number of samples in each treatment group:

```{r, results = "asis"}
# print the number of samples in each treatment group
n_samples_by_treatment <- metadata %>%
  group_by(treatment) %>%
  dplyr::summarise(n_samples = n())

base::cat(base::paste0("- ", n_samples_by_treatment$treatment, ": ", n_samples_by_treatment$n_samples, "\n"))
```

## limma/voom

### Filtering settings:

```{r, results = "asis"}
# print the filtering options the user has chosen to use
base::cat(base::paste0("- Min total count:", config$min_total_count, "\n", " - Min count:", config$min_count, "\n"))
```

```{r, limma_voom}
# create DGEList objects for the datasets (loop over all count datasets)
dge <- base::lapply(count_datasets, function(x) {
  
  edgeR::DGEList(counts = x, group = ordered_treatments)
  
})

# filter lowly expressed RNA
keep <- base::lapply(dge, function(x) {
  
  edgeR::filterByExpr(x, group=ordered_treatments, min.count = config$min_count, min.total.count = config$min_total_count)

})
```

```{r, eval = config$mirna_smrnaseq}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_mirna_smrnaseq_filtered <- dge$mirna_smrnaseq[keep$mirna_smrnaseq,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r, eval = config$mirna_excerpt}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_mirna_excerpt_filtered <- dge$mirna_excerpt[keep$mirna_excerpt,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r, eval = config$pirna_excerpt}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_pirna_excerpt_filtered <- dge$pirna_excerpt[keep$pirna_excerpt,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r, eval = config$trna_excerpt}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_trna_excerpt_filtered <- dge$trna_excerpt[keep$trna_excerpt,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r, eval = config$circrna_excerpt}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_circrna_excerpt_filtered <- dge$circrna_excerpt[keep$circrna_excerpt,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r, eval = config$gencode_excerpt}
# filtering and calulating normalisation factors (I couldn't get this to work in a list of datasets)
# (note. calcNormFactors doesn’t normalize the data, it just calculates normalization factors for use downstream)
dge_gencode_excerpt_filtered <- dge$gencode_excerpt[keep$gencode_excerpt,, keep.lib.sizes=FALSE] %>%
  edgeR::calcNormFactors(method="TMM")
```

```{r}
# create a model matrix
# also simplify the column names in the model matrix (loop over all count datasets)
model_matrix <- base::lapply(dge, function(x) {
  
  stats::model.matrix(~0+ordered_treatments) %>%
    base::as.data.frame() %>%
    dplyr::rename_with(~ (base::gsub("ordered_treatments", "", .x, fixed = TRUE))) %>%
    base::data.matrix()
  
})

# specify which groups to compare (based on the users choice of contrasts to compare) (loop over all count datasets)
contr_matrix <- base::lapply(model_matrix, function(x) {
  
  limma::makeContrasts(contrasts = c(config$contrasts), levels = base::colnames(x))
  
})
```

### Mean variance plot {.tabset .tabset-fade}

```{r, eval = config$mirna_smrnaseq, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = config$mirna_smrnaseq, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
mirna_smrnaseq_limmavoom_v <- limma::voom(dge_mirna_smrnaseq_filtered,  model_matrix$mirna_smrnaseq, plot=TRUE)
```

```{r, eval = config$mirna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = config$mirna_excerpt, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
mirna_excerpt_limmavoom_v <- limma::voom(dge_mirna_excerpt_filtered,  model_matrix$mirna_excerpt, plot=TRUE)
```

```{r, eval = config$pirna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = config$pirna_excerpt, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
pirna_excerpt_limmavoom_v <- limma::voom(dge_pirna_excerpt_filtered,  model_matrix$pirna_excerpt, plot=TRUE)
```

```{r, eval = config$trna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = config$trna_excerpt, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
trna_excerpt_limmavoom_v <- limma::voom(dge_trna_excerpt_filtered,  model_matrix$trna_excerpt, plot=TRUE)
```

```{r, eval = config$circrna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = config$circrna_excerpt, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
circrna_excerpt_limmavoom_v <- limma::voom(dge_circrna_excerpt_filtered,  model_matrix$circrna_excerpt, plot=TRUE)
```

```{r, eval = config$gencode_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = config$gencode_excerpt, out.width = "100%"}
# remove heteroscedascity from count data and plot mean variance
gencode_excerpt_limmavoom_v <- limma::voom(dge_gencode_excerpt_filtered,  model_matrix$gencodes_excerpt, plot=TRUE)
```

### Residual variances {.tabset .tabset-fade}

```{r, eval = config$mirna_smrnaseq, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = config$mirna_smrnaseq, out.width = "100%"}
# fit model to the expression values for each RNA
mirna_smrnaseq_limmavoom_vfit <- limma::lmFit(mirna_smrnaseq_limmavoom_v, model_matrix$mirna_smrnaseq)

# estimate contrast for each miRNA
mirna_smrnaseq_limmavoom_vfit <- limma::contrasts.fit(mirna_smrnaseq_limmavoom_vfit, contr_matrix$mirna_smrnaseq)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
mirna_smrnaseq_limmavoom_efit <- limma::eBayes(mirna_smrnaseq_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(mirna_smrnaseq_limmavoom_efit)
```

```{r, eval = config$mirna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = config$mirna_excerpt, out.width = "100%"}
# fit model to the expression values for each RNA
mirna_excerpt_limmavoom_vfit <- limma::lmFit(mirna_excerpt_limmavoom_v, model_matrix$mirna_excerpt)

# estimate contrast for each miRNA
mirna_excerpt_limmavoom_vfit <- limma::contrasts.fit(mirna_excerpt_limmavoom_vfit, contr_matrix$mirna_excerpt)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
mirna_excerpt_limmavoom_efit <- limma::eBayes(mirna_excerpt_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(mirna_excerpt_limmavoom_efit)
```

```{r, eval = config$pirna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = config$pirna_excerpt, out.width = "100%"}
# fit model to the expression values for each RNA
pirna_excerpt_limmavoom_vfit <- limma::lmFit(pirna_excerpt_limmavoom_v, model_matrix$pirna_excerpt)

# estimate contrast for each miRNA
pirna_excerpt_limmavoom_vfit <- limma::contrasts.fit(pirna_excerpt_limmavoom_vfit, contr_matrix$pirna_excerpt)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
pirna_excerpt_limmavoom_efit <- limma::eBayes(pirna_excerpt_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(pirna_excerpt_limmavoom_efit)
```

```{r, eval = config$trna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = config$trna_excerpt, out.width = "100%"}
# fit model to the expression values for each RNA
trna_excerpt_limmavoom_vfit <- limma::lmFit(trna_excerpt_limmavoom_v, model_matrix$trna_excerpt)

# estimate contrast for each miRNA
trna_excerpt_limmavoom_vfit <- limma::contrasts.fit(trna_excerpt_limmavoom_vfit, contr_matrix$trna_excerpt)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
trna_excerpt_limmavoom_efit <- limma::eBayes(trna_excerpt_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(trna_excerpt_limmavoom_efit)
```

```{r, eval = config$circrna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = config$circrna_excerpt, out.width = "100%"}
# fit model to the expression values for each RNA
circrna_excerpt_limmavoom_vfit <- limma::lmFit(circrna_excerpt_limmavoom_v, model_matrix$circrna_excerpt)

# estimate contrast for each miRNA
circrna_excerpt_limmavoom_vfit <- limma::contrasts.fit(circrna_excerpt_limmavoom_vfit, contr_matrix$circrna_excerpt)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
circrna_excerpt_limmavoom_efit <- limma::eBayes(circrna_excerpt_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(circrna_excerpt_limmavoom_efit)
```

```{r, eval = config$gencode_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = config$gencode_excerpt, out.width = "100%"}
# fit model to the expression values for each RNA
gencode_excerpt_limmavoom_vfit <- limma::lmFit(gencode_excerpt_limmavoom_v, model_matrix$gencode_excerpt)

# estimate contrast for each miRNA
gencode_excerpt_limmavoom_vfit <- limma::contrasts.fit(gencode_excerpt_limmavoom_vfit, contr_matrix$gencode_excerpt)

# empirical Bayes moderation - obtain more precise estimates of miRNA-wise variability
gencode_excerpt_limmavoom_efit <- limma::eBayes(gencode_excerpt_limmavoom_vfit)

# plot residual variances (code adapted from the "limma::plotSA()" function)
plotSA_interactive(gencode_excerpt_limmavoom_efit)
```

## DESeq2

```{r, deseq, results = "hide"}
# convert sample column to rownames since deseq seems to require this format
ordered_treatments <- treatments %>%
  tibble::remove_rownames() %>%
  tibble::column_to_rownames(var = "sample")

# create a DESeqDataSet object from my cleaned matrix and carry out deseq differential expression analysis (loop over all count datasets)
deseq <- base::lapply(count_datasets, function(x) {
  
  x %>%
    DESeq2::DESeqDataSetFromMatrix(colData = ordered_treatments,
                                   design = ~ treatment) %>%
    DESeq()
  
})
```

```{r, eval = contrasts[1]}
# comparison/contrast 1
# extract the results of the deseq differential expression analysis (loop over all count datasets)
deseq_results_1 <- base::lapply(deseq, function(x) {
  
  x %>%
    DESeq2::results(contrast = c("treatment", base::sub(" - .*", "", config$contrasts[1]), base::sub(".* - ", "", config$contrasts[1])),
                    lfcThreshold = min_logfc,
                    alpha = config$pvalue_cutoff)
  
})
```

```{r, eval = contrasts[2]}
# comparison/contrast 2
# extract the results of the deseq differential expression analysis (loop over all count datasets)
deseq_results_2 <- base::lapply(deseq, function(x) {
  
  x %>%
    DESeq2::results(contrast = c("treatment", base::sub(" - .*", "", config$contrasts[2]), base::sub(".* - ", "", config$contrasts[2])),
                    lfcThreshold = min_logfc,
                    alpha = config$pvalue_cutoff)
  
})
```

```{r, eval = contrasts[3]}
# comparison/contrast 3
# extract the results of the deseq differential expression analysis (loop over all count datasets)
deseq_results_3 <- base::lapply(deseq, function(x) {
  
  x %>%
    DESeq2::results(contrast = c("treatment", base::sub(" - .*", "", config$contrasts[3]), base::sub(".* - ", "", config$contrasts[3])),
                    lfcThreshold = min_logfc,
                    alpha = config$pvalue_cutoff)
  
})
```

```{r, eval = contrasts[4]}
# comparison/contrast 4
# extract the results of the deseq differential expression analysis (loop over all count datasets)
deseq_results_4 <- base::lapply(deseq, function(x) {
  
  x %>%
    DESeq2::results(contrast = c("treatment", base::sub(" - .*", "", config$contrasts[4]), base::sub(".* - ", "", config$contrasts[4])),
                    lfcThreshold = min_logfc,
                    alpha = config$pvalue_cutoff)
  
})
```

```{r, eval = contrasts[5]}
# comparison/contrast 5
# extract the results of the deseq differential expression analysis (loop over all count datasets)
deseq_results_5 <- base::lapply(deseq, function(x) {
  
  x %>%
    DESeq2::results(contrast = c("treatment", base::sub(" - .*", "", config$contrasts[5]), base::sub(".* - ", "", config$contrasts[5])),
                    lfcThreshold = min_logfc,
                    alpha = config$pvalue_cutoff)
  
})
```

```{r, eval = contrasts[6]}
# comparison/contrast 6
# extract the results of the deseq differential expression analysis (loop over all count datasets)
deseq_results_6 <- base::lapply(deseq, function(x) {
  
  x %>%
    DESeq2::results(contrast = c("treatment", base::sub(" - .*", "", config$contrasts[6]), base::sub(".* - ", "", config$contrasts[6])),
                    lfcThreshold = min_logfc,
                    alpha = config$pvalue_cutoff)
  
})
```

### Count outlier detection {.tabset .tabset-fade}

```{r, eval = config$mirna_smrnaseq, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = config$mirna_smrnaseq, out.width = "100%"}
# plot
data <- deseq_results_1$mirna_smrnaseq
W <- data$stat
maxCooks <- base::apply(SummarizedExperiment::assays(deseq$mirna_smrnaseq)[["cooks"]], 1, base::max)
idx <- !is.na(W)
base::plot(rank(W[idx]), maxCooks[idx], xlab = "rank of Wald statistic", 
           ylab = "maximum Cook's distance per RNA",
           ylim = c(0, 5), cex = .4, col = rgb(0, 0, 0, .3))
m <- base::ncol(deseq$mirna_smrnaseq)
p <- 3
graphics::abline(h = qf(.99, p, m - p))
```

```{r, eval = config$mirna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = config$mirna_excerpt, out.width = "100%"}
# plot
data <- deseq_results_1$mirna_excerpt
W <- data$stat
maxCooks <- base::apply(SummarizedExperiment::assays(deseq$mirna_excerpt)[["cooks"]], 1, base::max)
idx <- !is.na(W)
base::plot(rank(W[idx]), maxCooks[idx], xlab = "rank of Wald statistic", 
           ylab = "maximum Cook's distance per RNA",
           ylim = c(0, 5), cex = .4, col = rgb(0, 0, 0, .3))
m <- base::ncol(deseq$mirna_excerpt)
p <- 3
graphics::abline(h = qf(.99, p, m - p))
```

```{r, eval = config$pirna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = config$pirna_excerpt, out.width = "100%"}
# plot
data <- deseq_results_1$pirna_excerpt
W <- data$stat
maxCooks <- base::apply(SummarizedExperiment::assays(deseq$pirna_excerpt)[["cooks"]], 1, base::max)
idx <- !is.na(W)
base::plot(rank(W[idx]), maxCooks[idx], xlab = "rank of Wald statistic", 
           ylab = "maximum Cook's distance per RNA",
           ylim = c(0, 5), cex = .4, col = rgb(0, 0, 0, .3))
m <- base::ncol(deseq$pirna_excerpt)
p <- 3
graphics::abline(h = qf(.99, p, m - p))
```

```{r, eval = config$trna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = config$trna_excerpt, out.width = "100%"}
# plot
data <- deseq_results_1$trna_excerpt
W <- data$stat
maxCooks <- base::apply(SummarizedExperiment::assays(deseq$trna_excerpt)[["cooks"]], 1, base::max)
idx <- !is.na(W)
base::plot(rank(W[idx]), maxCooks[idx], xlab = "rank of Wald statistic", 
           ylab = "maximum Cook's distance per RNA",
           ylim = c(0, 5), cex = .4, col = rgb(0, 0, 0, .3))
m <- base::ncol(deseq$trna_excerpt)
p <- 3
graphics::abline(h = qf(.99, p, m - p))
```

```{r, eval = config$circrna_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = config$circrna_excerpt, out.width = "100%"}
# plot
data <- deseq_results_1$circrna_excerpt
W <- data$stat
maxCooks <- base::apply(SummarizedExperiment::assays(deseq$circrna_excerpt)[["cooks"]], 1, base::max)
idx <- !is.na(W)
base::plot(rank(W[idx]), maxCooks[idx], xlab = "rank of Wald statistic", 
           ylab = "maximum Cook's distance per RNA",
           ylim = c(0, 5), cex = .4, col = rgb(0, 0, 0, .3))
m <- base::ncol(deseq$circrna_excerpt)
p <- 3
graphics::abline(h = qf(.99, p, m - p))
```

```{r, eval = config$gencode_excerpt, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = config$gencode_excerpt, out.width = "100%"}
# plot
data <- deseq_results_1$gencode_excerpt
W <- data$stat
maxCooks <- base::apply(SummarizedExperiment::assays(deseq$gencode_excerpt)[["cooks"]], 1, base::max)
idx <- !is.na(W)
base::plot(rank(W[idx]), maxCooks[idx], xlab = "rank of Wald statistic", 
           ylab = "maximum Cook's distance per RNA",
           ylim = c(0, 5), cex = .4, col = rgb(0, 0, 0, .3))
m <- base::ncol(deseq$gencode_excerpt)
p <- 3
graphics::abline(h = qf(.99, p, m - p))
```

### MA plots {.tabset .tabset-fade}

```{r, eval = contrasts_chunk_eval$contrast_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("####", config$contrasts[[1]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_1, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_1$mirna_smrnaseq,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$mirna_excerpt_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_1, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_1$mirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$pirna_excerpt_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_1, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_1$pirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$trna_excerpt_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_1, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_1$trna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$circrna_excerpt_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_1, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_1$circrna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$gencode_excerpt_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_1, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_1$gencode_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = contrasts_chunk_eval$contrast_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("####", config$contrasts[[2]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_2, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_2$mirna_smrnaseq,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$mirna_excerpt_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_2, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_2$mirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$pirna_excerpt_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_2, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_2$pirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$trna_excerpt_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_2, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_2$trna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$circrna_excerpt_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_2, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_2$circrna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$gencode_excerpt_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_2, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_2$gencode_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = contrasts_chunk_eval$contrast_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("####", config$contrasts[[3]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_3, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_3$mirna_smrnaseq,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$mirna_excerpt_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_3, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_3$mirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$pirna_excerpt_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_3, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_3$pirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$trna_excerpt_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_3, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_3$trna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$circrna_excerpt_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_3, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_3$circrna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$gencode_excerpt_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_3, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_3$gencode_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = contrasts_chunk_eval$contrast_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("####", config$contrasts[[4]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_4, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_4$mirna_smrnaseq,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$mirna_excerpt_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_4, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_4$mirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$pirna_excerpt_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_4, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_4$pirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$trna_excerpt_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_4, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_4$trna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$circrna_excerpt_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_4, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_4$circrna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$gencode_excerpt_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_4, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_4$gencode_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = contrasts_chunk_eval$contrast_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("####", config$contrasts[[5]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_5, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_5$mirna_smrnaseq,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$mirna_excerpt_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_5, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_5$mirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$pirna_excerpt_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_5, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_5$pirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$trna_excerpt_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_5, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_5$trna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$circrna_excerpt_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_5, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_5$circrna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$gencode_excerpt_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_5, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_5$gencode_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = contrasts_chunk_eval$contrast_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("####", config$contrasts[[6]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_6, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_6$mirna_smrnaseq,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$mirna_excerpt_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_6, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_6$mirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$pirna_excerpt_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_6, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_6$pirna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$trna_excerpt_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_6, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_6$trna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$circrna_excerpt_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_6, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_6$circrna_excerpt,
               alpha = config$pvalue_cutoff)
```

```{r, eval = to_eval_chunk$gencode_excerpt_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("##### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_6, out.width = "100%"}
# plot
DESeq2::plotMA(deseq_results_6$gencode_excerpt,
               alpha = config$pvalue_cutoff)
```

## All differential expression results

```{r}
# make a list of all the efit objects in the global environment
efit <- base::do.call("list", base::mget(base::grep("_efit", base::names(.GlobalEnv), value=TRUE)))
```

```{r, eval = contrasts[1]}
# extract the results of the limma/voom differential expression analysis (first contrast) (loop over all efit datasets)
limma_voom_results_1 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 1,
                    n = Inf,
                    p.value = config$pvalue_cutoff,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_1 <- base::do.call(rbind, limma_voom_results_1)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_1 <- limma_voom_results_1 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_1 <- limma_voom_results_1 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_1 <- limma_voom_results_1 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)
```

```{r, eval = contrasts[2]}
# extract the results of the limma/voom differential expression analysis (second contrast) (loop over all efit datasets)
limma_voom_results_2 <- base::lapply(efit, function(x) {
  
  mirna_excerpt_limmavoom_efit %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 2,
                    n = Inf,
                    p.value = config$pvalue_cutoff,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_2 <- base::do.call(rbind, limma_voom_results_2)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_2 <- limma_voom_results_2 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[2]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_2 <- limma_voom_results_2 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_2 <- limma_voom_results_2 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)
```

```{r, eval = contrasts[3]}
# extract the results of the limma/voom differential expression analysis (third contrast) (loop over all efit datasets)
limma_voom_results_3 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 3,
                    n = Inf,
                    p.value = config$pvalue_cutoff,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_3 <- base::do.call(rbind, limma_voom_results_3)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_3 <- limma_voom_results_3 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[3]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_3 <- limma_voom_results_3 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_3 <- limma_voom_results_3 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)
```

```{r, eval = contrasts[4]}
# extract the results of the limma/voom differential expression analysis (fourth contrast) (loop over all efit datasets)
limma_voom_results_4 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 4,
                    n = Inf,
                    p.value = config$pvalue_cutoff,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_4 <- base::do.call(rbind, limma_voom_results_4)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_4 <- limma_voom_results_4 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[4]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_4 <- limma_voom_results_4 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_4 <- limma_voom_results_4 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)
```

```{r, eval = contrasts[5]}
# make a list of all the efit objects in the global environment
efit <- base::do.call("list", base::mget(base::grep("_efit", base::names(.GlobalEnv), value=TRUE)))

# extract the results of the limma/voom differential expression analysis (fifth contrast) (loop over all efit datasets)
limma_voom_results_5 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 5,
                    n = Inf,
                    p.value = config$pvalue_cutoff,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_5 <- base::do.call(rbind, limma_voom_results_5)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_5 <- limma_voom_results_5 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[5]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_5 <- limma_voom_results_5 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_5 <- limma_voom_results_5 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)
```

```{r, eval = contrasts[6]}
# extract the results of the limma/voom differential expression analysis (sixth contrast) (loop over all efit datasets)
limma_voom_results_6 <- base::lapply(efit, function(x) {
  
  x %>%
    limma::treat(lfc = min_logfc) %>%
    limma::topTreat(coef = 6,
                    n = Inf,
                    p.value = config$pvalue_cutoff,
                    adjust.method = "BH") %>%
    tibble::rownames_to_column("rna")
  
})

# collapse the list of dataframes into one
limma_voom_results_6 <- base::do.call(rbind, limma_voom_results_6)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
limma_voom_results_6 <- limma_voom_results_6 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("_limmavoom_efit.*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[6]) %>%
  dplyr::mutate(diff_expr_method = "limma/voom") %>%
  select(-source)

# select only some columns we're interested in
limma_voom_results_6 <- limma_voom_results_6 %>%
  select(rna, P.Value, adj.P.Val, logFC, rna_species, pipeline, comparison, diff_expr_method)

# rename adjusted p-value column before join
limma_voom_results_6 <- limma_voom_results_6 %>%
  dplyr::rename(p_value = P.Value) %>%
  dplyr::rename(adj_p_value = adj.P.Val) %>%
  dplyr::rename(log_fc = logFC)
```

```{r, eval = contrasts[1]}
# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames) (loop over all datasets)
deseq_results_1 <- base::lapply(deseq_results_1, function(x) {
  
  x %>%
    base::as.data.frame() %>%
    tibble::rownames_to_column("rna") 
  
})

# get the significant RNA's (that meet a logfc and p-value threshold)...
# ...for positive logFC
deseq_positive_1 <- base::lapply(deseq_results_1, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange > min_logfc), ]
  
})

# ...for negative logFC
deseq_negative_1 <- base::lapply(deseq_results_1, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange < base::paste0("-", min_logfc)), ]
  
})

# collapse the list of dataframes into one
deseq_positive_1 <- base::do.call(rbind, deseq_positive_1)
deseq_negative_1 <- base::do.call(rbind, deseq_negative_1)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
deseq_negative_1 <- deseq_negative_1 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

deseq_positive_1 <- deseq_positive_1 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

# bind negative and positive logFC results together
deseq_results_1 <- rbind(deseq_negative_1, deseq_positive_1)

# select only some columns we're interested in
deseq_results_1 <- deseq_results_1 %>%
  dplyr::select(rna, pvalue, padj, log2FoldChange, rna_species, pipeline, comparison, diff_expr_method)

# rename columns before join
deseq_results_1 <- deseq_results_1 %>%
  dplyr::rename(p_value = pvalue) %>%
  dplyr::rename(adj_p_value = padj) %>%
  dplyr::rename(log_fc = log2FoldChange)
```

```{r, eval = contrasts[2]}
# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames) (loop over all datasets)
deseq_results_2 <- base::lapply(deseq_results_2, function(x) {
  
  x %>%
    base::as.data.frame() %>%
    tibble::rownames_to_column("rna") 
  
})

# get the significant RNA's (that meet a logfc and p-value threshold)...
# ...for positive logFC
deseq_positive_2 <- base::lapply(deseq_results_2, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange > min_logfc), ]
  
})

# ...for negative logFC
deseq_negative_2 <- base::lapply(deseq_results_2, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange < base::paste0("-", min_logfc)), ]
  
})

# collapse the list of dataframes into one
deseq_positive_2 <- base::do.call(rbind, deseq_positive_2)
deseq_negative_2 <- base::do.call(rbind, deseq_negative_2)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
deseq_negative_2 <- deseq_negative_2 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

deseq_positive_2 <- deseq_positive_2 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

# bind negative and positive logFC results together
deseq_results_2 <- rbind(deseq_negative_2, deseq_positive_2)

# select only some columns we're interested in
deseq_results_2 <- deseq_results_2 %>%
  dplyr::select(rna, pvalue, padj, log2FoldChange, rna_species, pipeline, comparison, diff_expr_method)

# rename columns before join
deseq_results_2 <- deseq_results_2 %>%
  dplyr::rename(p_value = pvalue) %>%
  dplyr::rename(adj_p_value = padj) %>%
  dplyr::rename(log_fc = log2FoldChange)
```

```{r, eval = contrasts[3]}
# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames) (loop over all datasets)
deseq_results_3 <- base::lapply(deseq_results_3, function(x) {
  
  x %>%
    base::as.data.frame() %>%
    tibble::rownames_to_column("rna") 
  
})

# get the significant RNA's (that meet a logfc and p-value threshold)...
# ...for positive logFC
deseq_positive_3 <- base::lapply(deseq_results_3, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange > min_logfc), ]
  
})

# ...for negative logFC
deseq_negative_3 <- base::lapply(deseq_results_3, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange < base::paste0("-", min_logfc)), ]
  
})

# collapse the list of dataframes into one
deseq_positive_3 <- base::do.call(rbind, deseq_positive_3)
deseq_negative_3 <- base::do.call(rbind, deseq_negative_3)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
deseq_negative_3 <- deseq_negative_3 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

deseq_positive_3 <- deseq_positive_3 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

# bind negative and positive logFC results together
deseq_results_3 <- rbind(deseq_negative_3, deseq_positive_3)

# select only some columns we're interested in
deseq_results_3 <- deseq_results_3 %>%
  dplyr::select(rna, pvalue, padj, log2FoldChange, rna_species, pipeline, comparison, diff_expr_method)

# rename columns before join
deseq_results_3 <- deseq_results_3 %>%
  dplyr::rename(p_value = pvalue) %>%
  dplyr::rename(adj_p_value = padj) %>%
  dplyr::rename(log_fc = log2FoldChange)
```

```{r, eval = contrasts[4]}
# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames) (loop over all datasets)
deseq_results_4 <- base::lapply(deseq_results_4, function(x) {
  
  x %>%
    base::as.data.frame() %>%
    tibble::rownames_to_column("rna") 
  
})

# get the significant RNA's (that meet a logfc and p-value threshold)...
# ...for positive logFC
deseq_positive_4 <- base::lapply(deseq_results_4, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange > min_logfc), ]
  
})

# ...for negative logFC
deseq_negative_4 <- base::lapply(deseq_results_4, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange < base::paste0("-", min_logfc)), ]
  
})

# collapse the list of dataframes into one
deseq_positive_4 <- base::do.call(rbind, deseq_positive_4)
deseq_negative_4 <- base::do.call(rbind, deseq_negative_4)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
deseq_negative_4 <- deseq_negative_4 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

deseq_positive_4 <- deseq_positive_4 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

# bind negative and positive logFC results together
deseq_results_4 <- rbind(deseq_negative_4, deseq_positive_4)

# select only some columns we're interested in
deseq_results_4 <- deseq_results_4 %>%
  dplyr::select(rna, pvalue, padj, log2FoldChange, rna_species, pipeline, comparison, diff_expr_method)

# rename columns before join
deseq_results_4 <- deseq_results_4 %>%
  dplyr::rename(p_value = pvalue) %>%
  dplyr::rename(adj_p_value = padj) %>%
  dplyr::rename(log_fc = log2FoldChange)
```

```{r, eval = contrasts[5]}
# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames) (loop over all datasets)
deseq_results_5 <- base::lapply(deseq_results_5, function(x) {
  
  x %>%
    base::as.data.frame() %>%
    tibble::rownames_to_column("rna") 
  
})

# get the significant RNA's (that meet a logfc and p-value threshold)...
# ...for positive logFC
deseq_positive_5 <- base::lapply(deseq_results_5, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange > min_logfc), ]
  
})

# ...for negative logFC
deseq_negative_5 <- base::lapply(deseq_results_5, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange < base::paste0("-", min_logfc)), ]
  
})

# collapse the list of dataframes into one
deseq_positive_5 <- base::do.call(rbind, deseq_positive_5)
deseq_negative_5 <- base::do.call(rbind, deseq_negative_5)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
deseq_negative_5 <- deseq_negative_5 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

deseq_positive_5 <- deseq_positive_5 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

# bind negative and positive logFC results together
deseq_results_5 <- rbind(deseq_negative_5, deseq_positive_5)

# select only some columns we're interested in
deseq_results_5 <- deseq_results_5 %>%
  dplyr::select(rna, pvalue, padj, log2FoldChange, rna_species, pipeline, comparison, diff_expr_method)

# rename columns before join
deseq_results_5 <- deseq_results_5 %>%
  dplyr::rename(p_value = pvalue) %>%
  dplyr::rename(adj_p_value = padj) %>%
  dplyr::rename(log_fc = log2FoldChange)
```

```{r, eval = contrasts[6]}
# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames) (loop over all datasets)
deseq_results_6 <- base::lapply(deseq_results_6, function(x) {
  
  x %>%
    base::as.data.frame() %>%
    tibble::rownames_to_column("rna") 
  
})

# get the significant RNA's (that meet a logfc and p-value threshold)...
# ...for positive logFC
deseq_positive_6 <- base::lapply(deseq_results_6, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange > min_logfc), ]
  
})

# ...for negative logFC
deseq_negative_6 <- base::lapply(deseq_results_6, function(x) {
  
  x[ which(x$padj < config$pvalue_cutoff & x$log2FoldChange < base::paste0("-", min_logfc)), ]
  
})

# collapse the list of dataframes into one
deseq_positive_6 <- base::do.call(rbind, deseq_positive_6)
deseq_negative_6 <- base::do.call(rbind, deseq_negative_6)

# define which pipeline, differential expression method, rna species and comparison/contrast that was used (info stored in rownames)
deseq_negative_6 <- deseq_negative_6 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

deseq_positive_6 <- deseq_positive_6 %>%
  tibble::rownames_to_column("source") %>%
  dplyr::mutate(source = gsub("\\..*", "", source)) %>%
  dplyr::mutate(rna_species = gsub("_.*", "", source)) %>%
  dplyr::mutate(pipeline = gsub(".*_", "", source)) %>%
  dplyr::mutate(comparison = config$contrasts[1]) %>%
  dplyr::mutate(diff_expr_method = "DESeq2") %>%
  select(-source)

# bind negative and positive logFC results together
deseq_results_6 <- rbind(deseq_negative_6, deseq_positive_6)

# select only some columns we're interested in
deseq_results_6 <- deseq_results_6 %>%
  dplyr::select(rna, pvalue, padj, log2FoldChange, rna_species, pipeline, comparison, diff_expr_method)

# rename columns before join
deseq_results_6 <- deseq_results_6 %>%
  dplyr::rename(p_value = pvalue) %>%
  dplyr::rename(adj_p_value = padj) %>%
  dplyr::rename(log_fc = log2FoldChange)
```

```{r, eval = contrasts[1]}
# collapse results from deseq and limma/voom differential expression analysis together
diff_expr_results_1 <- dplyr::full_join(limma_voom_results_1, deseq_results_1, by = c("rna", "p_value", "adj_p_value", "log_fc", "rna_species", "pipeline", "comparison", "diff_expr_method"))
```

```{r, eval = contrasts[2]}
# collapse results from deseq and limma/voom differential expression analysis together
diff_expr_results_2 <- dplyr::full_join(limma_voom_results_2, deseq_results_2, by = c("rna", "p_value", "adj_p_value", "log_fc", "rna_species", "pipeline", "comparison", "diff_expr_method"))
```

```{r, eval = contrasts[3]}
# collapse results from deseq and limma/voom differential expression analysis together
diff_expr_results_3 <- dplyr::full_join(limma_voom_results_3, deseq_results_3, by = c("rna", "p_value", "adj_p_value", "log_fc", "rna_species", "pipeline", "comparison", "diff_expr_method"))
```

```{r, eval = contrasts[4]}
# collapse results from deseq and limma/voom differential expression analysis together
diff_expr_results_4 <- dplyr::full_join(limma_voom_results_4, deseq_results_4, by = c("rna", "p_value", "adj_p_value", "log_fc", "rna_species", "pipeline", "comparison", "diff_expr_method"))
```

```{r, eval = contrasts[5]}
# collapse results from deseq and limma/voom differential expression analysis together
diff_expr_results_5 <- dplyr::full_join(limma_voom_results_5, deseq_results_5, by = c("rna", "p_value", "adj_p_value", "log_fc", "rna_species", "pipeline", "comparison", "diff_expr_method"))
```

```{r, eval = contrasts[6]}
# collapse results from deseq and limma/voom differential expression analysis together
diff_expr_results_6 <- dplyr::full_join(limma_voom_results_6, deseq_results_6, by = c("rna", "p_value", "adj_p_value", "log_fc", "rna_species", "pipeline", "comparison", "diff_expr_method"))
```

*Note. only RNA's that were found to be significantly differentially expressed are included in this table*

```{r, all contrasts, out.width = "100%"}
# make a list of all the differential expression results objects in the global environment
diff_expr_results <- base::do.call("list", base::mget(base::grep("diff_expr_results_", base::names(.GlobalEnv), value=TRUE)))

# collapse all the differential expression resultsresults from all the contrasts analysed
diff_expr_results  <- base::do.call(rbind, diff_expr_results)

# create a column that defines if an RNA is significant or not (at three significance levels and based on
# both limma/voom and deseq results). I ordered the case_when conditions from not significant to significant
# at the 1% level. This way, if an observation is true in several of these case_when categories, it'll be 
# marked with the smallest p-value category is fits into, sort of "rounded up" to the smallest p-value,
# because this case_when function seems to overwrite a value if it falls in a later case_when category
# This data will be used as a "flag" (using highlighting) in the downstream app to indicate a possibly
# significantly differentiated RNA the user of the app might want to investigate
diff_expr_results <- diff_expr_results %>%
  dplyr::mutate(significance = dplyr::case_when(adj_p_value > 0.10 ~ "greater_than_10%",
                                                (adj_p_value > 0.05) & (adj_p_value <= 0.10) ~ "significant_10%",
                                                (adj_p_value > 0.01) & (adj_p_value <= 0.05) ~ "significant_5%",
                                                (adj_p_value <= 0.01) ~ "significant_1%"))

# create datatable to explore the data
DT::datatable(diff_expr_results %>%
                dplyr::select(rna,
                              rna_species,
                              pipeline,
                              diff_expr_method,
                              comparison,
                              log_fc,
                              p_value,
                              adj_p_value,
                              significance) %>%
                dplyr::mutate(across(c(rna_species,
                                       pipeline,
                                       diff_expr_method,
                                       comparison,
                                       significance), base::as.factor)) %>%
                dplyr::mutate(across(c(log_fc,
                                       p_value,
                                       adj_p_value), base::as.double)) %>%
                dplyr::mutate(across(c(p_value,
                                       adj_p_value), ~base::round(.x, digits = 6))) %>%
                dplyr::mutate(across(c(log_fc), ~base::round(.x, digits = 4))),
              filter = "top",
              rownames = FALSE,
              colnames = c("RNA",
                           "RNA species",
                           "Pipeline",
                           "Differential expression method",
                           "Comparison",
                           "Log fold change",
                           "p-value",
                           "Adjusted p-value",
                           "Significance"),
              extensions = base::list("ColReorder" = NULL,
                                      "Buttons" = NULL,
                                      "FixedColumns" = base::list(leftColumns=1)),
              options = list(
                dom = "BRrltpi",
                autoWidth = TRUE,
                buttons =
                  list("copy",
                       list(extend = "collection",
                            buttons = c("csv", "excel", "pdf"),
                            text = "Download"),
                       I("colvis"))))
```

```{r}
# write differential expression results to file
diff_expr_results %>%
  utils::write.table(file = "./diff_expr_results/diff_expr_results.tsv", row.names=FALSE, sep="\t")
```

```{r, cleanup, results = "hide"}
# clean up
rm(list = ls())
```
