---
title: "Differential expression MA plots - deseq2"
author:
  # - Jane Doe^[Institution Two, jane@example.org]      # add report authors (uncomment if using)
  # - John Doe^[Institution One, john@example.org]      # add a second report author (uncomment if using)
date: "Date: `r base::format(Sys.time(), '%d/%m/%Y')`"
always_allow_html: true
output:
  bookdown::html_document2:
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: true
    number_sections: false
    self_contained: yes
    css: ../esr-styles.css
    highlight: pygments
    includes:
      after_body: ../esr_footer.html
  editor_options: 
    chunk_output_type: console
---

<img src="../images/ESR_logo.svg" style="position:absolute;top:19px;right:40px;width:23%;" />

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, error = FALSE, message = FALSE, fig.align = "center")
```

```{r}
# load libraries
library(dplyr)
library(limma)
library(DESeq2)
library(janitor)
library(edgeR)
library(DT)
library(apeglm)
library(plotly)
library(heatmaply)
library(gtools)
library(textshape)
library(tidyr)

# for icons at bottom of doc
htmltools::tagList(rmarkdown::html_dependency_font_awesome())

# read in yaml config file
config <- yaml::yaml.load_file("../config/config.yaml")

# read in metadata
metadata <- utils::read.csv(base::file.path(config$metadata_path))

# read in deseq differential expression results
deseq_diff_expr_results <- utils::read.table("./diff_expr_results/all_deseq_diff_expr_results.tsv", header = TRUE)
```

## Info

General info:

- Differential expression analysis using [deseq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8).
- Some values have been rounded

Datasets analysed:

```{r, results = "asis"}
# print the datasets the user has chosen to analyse for this document
base::cat(base::paste0(" - mirna smrnaseq: ", config$mirna_smrnaseq, "\n",
                       " - mirna excerpt: ", config$mirna_excerpt, "\n",
                       " - pirna excerpt: ", config$pirna_excerpt, "\n",
                       " - trna excerpt: ", config$trna_excerpt, "\n",
                       " - circrna excerpt: ", config$circrna_excerpt, "\n",
                       " - gencode excerpt: ", config$gencode_excerpt, "\n"))
```

Treatment comparisons: 

```{r, results = "asis"}
# print the treatment group comparisons the user has chosen to analyse
base::cat(base::paste0("- ", config$contrasts, "\n"))
```

```{r, results = "asis"}
# print the number of samples analysed
base::cat(base::paste0("Total number of samples: ", base::length(base::unique(metadata$sample))))
```

Number of samples in each treatment group:

```{r, results = "asis"}
# print the number of samples in each treatment group
n_samples_by_treatment <- metadata %>%
  group_by(treatment) %>%
  dplyr::summarise(n_samples = n())

base::cat(base::paste0("- ", n_samples_by_treatment$treatment, ": ", n_samples_by_treatment$n_samples, "\n"))
```

```{r, results = "hide"}
# create a list defining which code chunks to analyse (that are set to TRUE for both the datasets and contrasts to analyse) based on the yaml user configuration file
to_eval_chunk <- config[c("mirna_smrnaseq",
                          "mirna_excerpt",
                          "pirna_excerpt",
                          "trna_excerpt",
                          "circrna_excerpt",
                          "gencode_excerpt")] %>%
  rep(times = 10)

names(to_eval_chunk)[1:6] <- paste(names(to_eval_chunk)[1:6], "1", sep = "_")
names(to_eval_chunk)[7:12] <- paste(names(to_eval_chunk)[7:12], "2", sep = "_")
names(to_eval_chunk)[13:18] <- paste(names(to_eval_chunk)[13:18], "3", sep = "_")
names(to_eval_chunk)[19:24] <- paste(names(to_eval_chunk)[19:24], "4", sep = "_")
names(to_eval_chunk)[25:30] <- paste(names(to_eval_chunk)[25:30], "5", sep = "_")
names(to_eval_chunk)[31:36] <- paste(names(to_eval_chunk)[31:36], "6", sep = "_")
names(to_eval_chunk)[37:42] <- paste(names(to_eval_chunk)[37:42], "7", sep = "_")
names(to_eval_chunk)[43:48] <- paste(names(to_eval_chunk)[43:48], "8", sep = "_")
names(to_eval_chunk)[49:54] <- paste(names(to_eval_chunk)[49:54], "9", sep = "_")
names(to_eval_chunk)[55:60] <- paste(names(to_eval_chunk)[55:60], "10", sep = "_")

n_contrasts <- base::length(config$contrasts)

to_eval_chunk <- if(n_contrasts == 1) {
  replace(to_eval_chunk, 7:60, FALSE)
} else if(n_contrasts == 2) {
  replace(to_eval_chunk, 13:60, FALSE)
} else if(n_contrasts == 3) {
  replace(to_eval_chunk, 19:60, FALSE)
} else if(n_contrasts == 4) {
  replace(to_eval_chunk, 25:60, FALSE)
} else if(n_contrasts == 5) {
  replace(to_eval_chunk, 31:60, FALSE)
} else if(n_contrasts == 6) {
  replace(to_eval_chunk, 37:60, FALSE)
} else if(n_contrasts == 7) {
  replace(to_eval_chunk, 43:60, FALSE)
} else if(n_contrasts == 8) {
  replace(to_eval_chunk, 49:60, FALSE)
} else if(n_contrasts == 9) {
  replace(to_eval_chunk, 55:60, FALSE)
} else if(n_contrasts == 10) {
  to_eval_chunk
}

# create a list defining which code chunks to analyse (based on the number of contrasts to analyse) based on the yaml user configuration file
contrasts_chunk_eval <- list(contrast_1 = TRUE,
                             contrast_2 = TRUE,
                             contrast_3 = TRUE,
                             contrast_4 = TRUE,
                             contrast_5 = TRUE,
                             contrast_6 = TRUE,
                             contrast_7 = TRUE,
                             contrast_8 = TRUE,
                             contrast_9 = TRUE,
                             contrast_10 = TRUE)

contrasts_chunk_eval <- if(n_contrasts == 1) {
  replace(contrasts_chunk_eval, 2:10, FALSE)
} else if(n_contrasts == 2) {
  replace(contrasts_chunk_eval, 3:10, FALSE)
} else if(n_contrasts == 3) {
  replace(contrasts_chunk_eval, 4:10, FALSE)
} else if(n_contrasts == 4) {
  replace(contrasts_chunk_eval, 5:10, FALSE)
} else if(n_contrasts == 5) {
  replace(contrasts_chunk_eval, 6:10, FALSE)
} else if(n_contrasts == 6) {
  replace(contrasts_chunk_eval, 7:10, FALSE)
} else if(n_contrasts == 7) {
  replace(contrasts_chunk_eval, 8:10, FALSE)
} else if(n_contrasts == 8) {
  replace(contrasts_chunk_eval, 9:10, FALSE)
} else if(n_contrasts == 9) {
  replace(contrasts_chunk_eval, 10, FALSE)
} else if(n_contrasts == 10) {
  contrasts_chunk_eval
}

# evalutate/setup minimum logFC threshold
min_logfc <- base::eval(base::parse(text = config$min_lfc))
```

```{r}
# setup a function to plot MA plots
MA_interactive <- function(data_frame) {
  
  # plot
  plotly::plot_ly(data = data_frame,
                  x = ~ baseMean,
                  y = ~ log_fc,
                  type = "scatter",
                  marker = list(color = "#0097db", opacity = 0.7),
                  hoverinfo = "text",
                  text = ~ base::paste(" RNA:", rna,
                                       "</br></br>",
                                       "Mean of normalized counts:", base::round(baseMean, digits = 0),
                                       "</br>",
                                       "Log fold change:", base::round(log_fc, digits = 4),
                                       "</br>",
                                       "Adjusted p-value:", base::round(adj_p_value, digits = 10)),
                  showlegend = FALSE) %>%
    
    # add significant points in red
    add_trace(data = dplyr::filter(data_frame, (adj_p_value <= 0.1) & ((log_fc >= min_logfc) | (log_fc <= -min_logfc))),
              x = ~ baseMean,
              y = ~ log_fc,
              type = "scatter",
              marker = list(color = "#ec1515", opacity = 0.7)) %>%
    
    # add significant points in yellow
    add_trace(data = dplyr::filter(data_frame, (adj_p_value <= 0.05) & ((log_fc >= min_logfc) | (log_fc <= -min_logfc))),
              x = ~ baseMean,
              y = ~ log_fc,
              type = "scatter",
              marker = list(color = "#febf2a", opacity = 0.7)) %>%
    
    # add significant points in green
    add_trace(data = dplyr::filter(data_frame, (adj_p_value <= 0.01) & ((log_fc >= min_logfc) | (log_fc <= -min_logfc))),
              x = ~ baseMean,
              y = ~ log_fc,
              type = "scatter",
              marker = list(color = "#85C659", opacity = 0.7)) %>%
    
    # formatting
    plotly::layout(xaxis = base::list(title = "Mean of normalized counts",
                                      type = "log"),
                   yaxis = base::list(title = "Log fold change"))
}
```

## MA plots {.tabset .tabset-fade}

```{r, results = "asis"}
base::cat("These plots plot the log2 fold changes versus the mean of normalized counts. These can be used to visualize the differences between treatment group comparisons. In general we would expect the expression of RNA's to remain consistent between treatment groups and so the MA plot should be similar to the shape of a trumpet with most points residing on a y intercept of 0. Differentially expressed RNA's are marked blue.")
```

```{r, results = "asis"}
base::cat(base::paste0("RNA's with a **p-value** smaller than 0.1 and a log fold change greater/smaller than +-", config$min_lfc, " are plotted here."))
```

<div class="alert alert-custom-blue">

```{r, results = "asis"}
base::cat(base::paste0("Blue: RNA's with an **adjusted p-value** smaller than 0.1 and a log fold change greater/smaller than +-", config$min_lfc))
```

</div>

<div class="alert alert-custom-red">

```{r, results = "asis"}
base::cat(base::paste0("Red: RNA's with an **adjusted p-value** smaller than 0.1 and a log fold change greater/smaller than +-", config$min_lfc))
```

</div>

<div class="alert alert-custom-yellow">

```{r, results = "asis"}
base::cat(base::paste0("Yellow: RNA's with an **adjusted p-value** smaller than 0.05 and a log fold change greater/smaller than +-", config$min_lfc))
```

</div>

<div class="alert alert-custom-green">

```{r, results = "asis"}
base::cat(base::paste0("Green: RNA's with an **adjusted p-value** smaller than 0.01 and a log fold change greater/smaller than +-", config$min_lfc))
```

</div>

```{r, eval = contrasts_chunk_eval$contrast_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("###", config$contrasts[[1]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_1, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[1]]) & (pipeline == "smrnaseq")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$mirna_excerpt_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_1, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[1]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$pirna_excerpt_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_1, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "pirna") & (comparison == config$contrasts[[1]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$trna_excerpt_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_1, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "trna") & (comparison == config$contrasts[[1]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$circrna_excerpt_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_1, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "circrna") & (comparison == config$contrasts[[1]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$gencode_excerpt_1, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_1, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "gencode") & (comparison == config$contrasts[[1]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = contrasts_chunk_eval$contrast_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("###", config$contrasts[[2]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_2, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[2]]) & (pipeline == "smrnaseq")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$mirna_excerpt_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_2, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[2]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$pirna_excerpt_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_2, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "pirna") & (comparison == config$contrasts[[2]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$trna_excerpt_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_2, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "trna") & (comparison == config$contrasts[[2]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$circrna_excerpt_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_2, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "circrna") & (comparison == config$contrasts[[2]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$gencode_excerpt_2, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_2, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "gencode") & (comparison == config$contrasts[[2]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = contrasts_chunk_eval$contrast_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("###", config$contrasts[[3]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_3, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[3]]) & (pipeline == "smrnaseq")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$mirna_excerpt_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_3, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[3]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$pirna_excerpt_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_3, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "pirna") & (comparison == config$contrasts[[3]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$trna_excerpt_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_3, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "trna") & (comparison == config$contrasts[[3]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$circrna_excerpt_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_3, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "circrna") & (comparison == config$contrasts[[3]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$gencode_excerpt_3, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_3, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "gencode") & (comparison == config$contrasts[[3]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = contrasts_chunk_eval$contrast_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("###", config$contrasts[[4]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_4, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[4]]) & (pipeline == "smrnaseq")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$mirna_excerpt_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_4, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[4]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$pirna_excerpt_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_4, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "pirna") & (comparison == config$contrasts[[4]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$trna_excerpt_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_4, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "trna") & (comparison == config$contrasts[[4]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$circrna_excerpt_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_4, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "circrna") & (comparison == config$contrasts[[4]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$gencode_excerpt_4, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_4, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "gencode") & (comparison == config$contrasts[[4]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = contrasts_chunk_eval$contrast_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("###", config$contrasts[[5]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_5, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[5]]) & (pipeline == "smrnaseq")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$mirna_excerpt_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_5, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[5]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$pirna_excerpt_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_5, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "pirna") & (comparison == config$contrasts[[5]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$trna_excerpt_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_5, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "trna") & (comparison == config$contrasts[[5]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$circrna_excerpt_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_5, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "circrna") & (comparison == config$contrasts[[5]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$gencode_excerpt_5, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_5, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "gencode") & (comparison == config$contrasts[[5]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = contrasts_chunk_eval$contrast_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("###", config$contrasts[[6]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_6, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[6]]) & (pipeline == "smrnaseq")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$mirna_excerpt_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_6, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[6]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$pirna_excerpt_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_6, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "pirna") & (comparison == config$contrasts[[6]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$trna_excerpt_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_6, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "trna") & (comparison == config$contrasts[[6]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$circrna_excerpt_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_6, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "circrna") & (comparison == config$contrasts[[6]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$gencode_excerpt_6, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_6, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "gencode") & (comparison == config$contrasts[[6]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = contrasts_chunk_eval$contrast_7, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("###", config$contrasts[[7]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_7, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_7, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[7]]) & (pipeline == "smrnaseq")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$mirna_excerpt_7, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_7, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[7]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$pirna_excerpt_7, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_7, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "pirna") & (comparison == config$contrasts[[7]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$trna_excerpt_7, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_7, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "trna") & (comparison == config$contrasts[[7]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$circrna_excerpt_7, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_7, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "circrna") & (comparison == config$contrasts[[7]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$gencode_excerpt_7, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_7, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "gencode") & (comparison == config$contrasts[[7]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = contrasts_chunk_eval$contrast_8, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("###", config$contrasts[[8]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_8, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_8, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[8]]) & (pipeline == "smrnaseq")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$mirna_excerpt_8, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_8, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[8]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$pirna_excerpt_8, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_8, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "pirna") & (comparison == config$contrasts[[8]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$trna_excerpt_8, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_8, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "trna") & (comparison == config$contrasts[[8]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$circrna_excerpt_8, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_8, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "circrna") & (comparison == config$contrasts[[8]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$gencode_excerpt_8, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_8, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "gencode") & (comparison == config$contrasts[[8]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = contrasts_chunk_eval$contrast_9, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("###", config$contrasts[[9]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_9, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_9, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[9]]) & (pipeline == "smrnaseq")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$mirna_excerpt_9, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_9, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[9]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$pirna_excerpt_9, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_9, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "pirna") & (comparison == config$contrasts[[9]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$trna_excerpt_9, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_9, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "trna") & (comparison == config$contrasts[[9]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$circrna_excerpt_9, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_9, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "circrna") & (comparison == config$contrasts[[9]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$gencode_excerpt_9, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_9, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "gencode") & (comparison == config$contrasts[[9]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = contrasts_chunk_eval$contrast_10, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("###", config$contrasts[[10]], "{.tabset .tabset-fade}")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_10, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (smrnaseq)")
```

```{r, eval = to_eval_chunk$mirna_smrnaseq_10, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[10]]) & (pipeline == "smrnaseq")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$mirna_excerpt_10, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### miRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$mirna_excerpt_10, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "mirna") & (comparison == config$contrasts[[10]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$pirna_excerpt_10, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### piRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$pirna_excerpt_10, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "pirna") & (comparison == config$contrasts[[10]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$trna_excerpt_10, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### tRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$trna_excerpt_10, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "trna") & (comparison == config$contrasts[[10]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$circrna_excerpt_10, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### circRNA's (excerpt)")
```

```{r, eval = to_eval_chunk$circrna_excerpt_10, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "circrna") & (comparison == config$contrasts[[10]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, eval = to_eval_chunk$gencode_excerpt_10, results = "asis"}
# optionally include/print markdown header for this section if being analysed
base::cat("#### gencodes (excerpt)")
```

```{r, eval = to_eval_chunk$gencode_excerpt_10, out.width = "100%"}
deseq_diff_expr_results %>%
  dplyr::filter((rna_species == "gencode") & (comparison == config$contrasts[[10]]) & (pipeline == "excerpt")) %>%
  MA_interactive()
```

```{r, cleanup, results = "hide"}
# clean up
rm(list = ls())
```
